MSC3952: Intentional Mentions (#3952)
matrix-org/matrix-spec-proposals: proposals/3952-intentional-mentions.md

# MSC3952: Intentional Mentions

Mentioning other users on Matrix is difficult -- it is not possible to know if
[mentioning a user by display name or Matrix ID](https://github.com/matrix-org/matrix-spec/issues/353)
will count as a mention, but is also too easy to mistakenly mention a user.

(Note that throughout this proposal "mention" is considered equivalent to a "ping"
or highlight notification.)

Some situations that result in unintentional mentions include:

* Replying to a message will re-issue pings from the initial message due to
  [fallback replies](https://spec.matrix.org/v1.5/client-server-api/#fallbacks-for-rich-replies). [^1]

> * Each time a message is edited the new version will be re-evaluated for mentions.

Author: Spurious notifications from message edits isn't fixed by this MSC.

Reviewer: I don't think that this should be in scope for this MSC as this can already be handled using existing push rules to skip notifying if it's an `m.replace`ment.

https://github.com/beeper/synapse/blame/beeper/rust/src/push/base_rules.rs#L98-L114

Author: Yep, I don't think this is in-scope for this MSC, this is mostly a note to myself that it is a bit odd to include it in the list here, but then not fix it.

Author: > I don't think that this should be in scope for this MSC as this can already be handled using existing push rules to skip notifying if it's an `m.replace`ment.
> 
> [beeper/synapse@`beeper`/rust/src/push/base_rules.rs#L98-L114 (blame)](https://github.com/beeper/synapse/blame/beeper/rust/src/push/base_rules.rs#L98-L114)

Although... would you mind if I wrote an MSC for this behavior? 😄

Author: > Yep, I don't think this is in-scope for this MSC, this is mostly a note to myself that it is a bit odd to include it in the list here, but then not fix it.

I added a note in a2896b2318812257971c854b306e3c870edc03e0.

Author: I think I might have originally included this since we would want to somehow define what happens if an edit modifies the `mentions` list. I suspect we should just not allow that.

Reviewer 2: > Although... would you mind if I wrote an MSC for this behavior? 😄

Do not mind at all, just thinking of making this a base push rule for everyone in the spec?

Author: > > Although... would you mind if I wrote an MSC for this behavior? 😄
> 
> Do not mind at all, just thinking of making this a base push rule for everyone in the spec?

Yes, pretty much. I'll re-use your unstable ID so you won't need to touch anything. 👍

Author: I started writing this and then realized edits interact weirdly with this MSC, so will probably define that first: https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1067414439

Author: (I still plan to do this, but it isn't blocking this MSC so I'm going to resolve the thread. I'll try to remember to cross-link when I write it.)

        "notify",
        {
            "set_tweak": "sound",
            "value": "default"
        },
        {
            "set_tweak": "highlight"
        }
    ]
}
```

The `is_mention` push condition is defined as[^3]:

> This matches messages where the `content.mentions` is an array containing the
> owner’s Matrix ID in the first 10 entries. This condition has no parameters.
> If the `mentions` key is absent or not an array then the rule does not match;
> any array entries which are not strings are ignored, but count against the limit.

An example matching event is provided below:

```json
{
    "content": {
        "body": "This is an example mention @alice:example.org",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>This is an example mention</b> <a href='https://matrix.to/#/@alice:example.org'>Alice</a>",
        "msgtype": "m.text",
        "mentions": ["@alice:example.org"]
    },
    "event_id": "$143273582443PhrSn:example.org",
    "origin_server_ts": 1432735824653,
    "room_id": "!somewhere:over.the.rainbow",
    "sender": "@example:example.org",
    "type": "m.room.message",
    "unsigned": {
        "age": 1234
    }
}
```

> ### Backwards compatibility

Author: Somewhere this should mention that users should feel free to use keywords if they would still like their display name (or localpart of their Matrix ID) to generate highlight notifications.

The the [`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules)
and [`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)
push rules are both deprecated. To avoid the unintentional mentions they are both
modified to only apply when the `mentions` field is missing. As this is for
backwards-compatibility it is not implemented using a generic mechanism, but

> behavior specific to those push rules with those IDs.

Author: I couldn't come up with another option unless there was a condition to check if a field was missing (but that seems a bit odd).

Also `.m.rule.contains_user_name` is a content rule, which can't have conditions.

Author: Another option would be to have clients start setting `m.mentions` field ASAP and leave these push rules in place for now (as-is). This wouldn't fix all the unintentional pings problems until these push rules are removed at a later date, however.

While this is being deployed there will be a mismatch for legacy clients which
implement the deprecated `.m.rule.contains_display_name` and `.m.rule.contains_user_name`
push rules locally while the `.m.rule.user_is_mentioned` push rule is used on
the homeserver; as messages which
[mention the user should also include the user's localpart](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
in the message `body` it is likely that the `.m.rule.contains_user_name`

> will also match. It is postulated that this will not cause issues in most cases.

Author: I'm not fully convinced this won't be an issue, will need to do some testing.

Author: @andybalaam Suggested that we just fully disable these rules when extensible events gets its own room version and (at that point) no longer require sending an empty `m.mentions` on each event.

Author: I added @andybalaam's suggestion in e346fbb27bf1374f0d58b38ea2d1b1694397c305.

## Potential issues

> ### Abuse potential

Reviewer 3: To add: it may also be turned into a feature, where a client displays "implicit mentions" / "CCs" on top or at the bottom of the message, de-emphasised, to show users who read that message the additional mentions that were added to it outside of the message body itself.

Author: Thanks! I added some bits about this at 99ab781f03bdea04cea34a418162d6968309cc13.

using a separate field for mentions:

* The number of mentions is trivially limited.
* Various forms of "mention bombing" are no longer possible.
* It is simpler to collect data on how many users are being mentioned (without
  having to process the textual `body` for every user's display name and local
  part).

Nonetheless, the conversations did result in some ideas to combat the potential
for abuse, many of which apply regardless of whether this proposal is implemented.

Clients could expose *why* an event has caused a notification and give users inline
tools to combat potential abuse. For example, a client might show a tooltip:

> The sender of the message (`@alice:example.org`) mentioned you in this event.
>
> Block `@alice:example.org` from sending you messages? `[Yes]` `[No]`

It could also be worth exposing user interface for moderators to see messages
which mention many users.

A future MSC might wish to explore features for trusted contacts or soft-ignores
to give users more control over who can generate notifications.

Overall the proposal does not seem to increase the potential for malicious behavior.

## Alternatives

> ### Prior proposals

Reviewer 2: Also https://github.com/matrix-org/matrix-spec-proposals/pull/1796 ?

This is my preferred proposal, as it allows server-side push rules to handle mentions, with the trade-off of leaking a bit more metadata

Reviewer: #1796 only suggests adding the `m.mentions` field to `m.room.encrypted` events so it would need to be expanded to unencrypted events as well to bring it to feature-parity with this proposal.

Author: I somehow did not come across that one in my research. It sounds awfully similar to this though. Will take a closer look.

Author: I read through this again and I think there's a few main changes (shout if I'm missing something):

1. Keeps the list of users unencrypted so servers can help out here.
2. Allow `@room` as a special value to also avoid issues with room notifications.
3. `m.room.encrypted` events vs. all events.

2 I think is fairly non-controversial and would make sense (I didn't find issues about edge-cases involving it, but I think you can extrapolate potential ones from the issues with user mentions).

For 1, as you highlight above, the tradeoff is mostly a bit more metadata leaking. I think this is probably worthwhile, although clients still need to decrypt if they want keyword notifications. Any chance you could write a quick blurb on what you think this tradeoff is useful? I'm not fully sure what clients win from this if they still need to decrypt for keywords.

And finally, I think expanding to all events is reasonable, especially with things like extensible events coming out.

Author: > Any chance you could write a quick blurb on what you think this tradeoff is useful? I'm not fully sure what clients win from this if they still need to decrypt for keywords.

What I'm trying to say here -- is the rationale given in MSC1796 still valid today or do most clients decrypt all events anyway?

Reviewer 2: The biggest use case for leaving mentions unencrypted is for allowing "muted-except-for-mentions" push rules for large rooms. I want a push when I get mentioned in a large encrypted room that we use for team communication but I don't want pushes for every message. With a platform like iOS that doesn't get to sync regularly in the background, the server needs enough information to decide whether to send a push to my mobile app or not, and today it doesn't.

Without this, yes the iOS app can sync events and decrypt them and run push rules against them when the app is foregrounded, but at that point you probably just want to update notifcation/highlight counts as opposed to notifying the user of a specific message from some time ago.

Reviewer 3: I'd ask to make it configurable between unencrypted and encrypted mentions, as it sounds like it could be a bit too much of a data leak.

Reviewer 2: Configurable makes sense to me, the user generating the event can make the choice where to put the mention data. This does add some complexity to everything else as now things like push rules would need to check two spots.

Hrm, maybe this is more complicated. Do clients run push rules on both encrypted and decrypted payloads? Is the resulting action the superset of both runs? Or do you take the higher priority rule? What if both runs result in different push rules in the same priority group being matched, one to supress and one to notify?

Or do you just special case the mention rule type as the only one that looks at both encrypted and decrypted content, and you evaluate both payloads at the same time as part of the same push rule "pass". This probably makes more sense.

Author: > Do clients run push rules on both encrypted and decrypted payloads? Is the resulting action the superset of both runs?

I think they run it only on the decrypted payload, although the decrypted payload includes the cleartext keys, I think? Clients processing push rules is poorly specced.

> Or do you just special case the mention rule type as the only one that looks at both encrypted and decrypted content, and you evaluate both payloads at the same time as part of the same push rule "pass". This probably makes more sense.

Not saying we cannot do this, but this would [be different behavior than `m.relates_to`](https://spec.matrix.org/v1.5/client-server-api/#forming-relationships-between-events) where the server and client behavior are purposefully required to be the same. My gut feeling is that we should follow something similar here.

Author: I added some info on this in 4de784893e53995991b18f44329ca074b54547ae and 5dfc0bce5793f761073ec1da12eb66fa0f4508ba, but it likely needs a few more pieces of info.

       }
     ]
  }
  ```

  </summary>

### Room version for backwards compatibility

Alternative backwards compatibility suggestions included using a new room version,
similar to [MSC3932](https://github.com/matrix-org/matrix-spec-proposals/pull/3932)
for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable
room version before use, which would unnecessarily delay usage.

## Security considerations

None foreseen.

## Unstable prefix

During development the `.org.matrix.msc3952.is_user_mentioned` push rule will be
used. If a client sees this rule available it should apply the custom logic discussed
in the [backwards compatibility](#backwards-compatibility) section.

## Dependencies

N/A

[^1]: This MSC does not attempt to solve this problem, but [MSC2781](https://github.com/matrix-org/matrix-spec-proposals/pull/2781)
proposes removing reply fallbacks which would solve it. Although as noted in
[MSC3676](https://github.com/matrix-org/matrix-spec-proposals/pull/3676) this
may require landing [MSC3664](https://github.com/matrix-org/matrix-doc/pull/3664)

> in order to receive notifications of replies.

Author: Maybe this MSC should recommend that clients add the user to the `mentions` array when replying? Otherwise we would lose the ability of replies to cause notifications.

Reviewer 4: Ensuring that users get a notification when their message has been replied to would be ideal (just like when a thread was started off your message in the main timeline).

Reviewer: I disagree with adding the replied-to-user in `mentions`.

Notifications for replies can already be handled by existing push rules once https://github.com/matrix-org/matrix-spec-proposals/pull/3664 is merged.

https://github.com/matrix-org/synapse/pull/11804 already has an implementation of the MSC: https://github.com/beeper/synapse/blob/beeper/rust/src/push/base_rules.rs#L181-L196

Author: > Notifications for replies can already be handled by existing push rules once #3664

I think this is pretty much what the paragraph I quoted says?

Author: The benefit of using the `mentions` array here is that it signals *sender* intention of whether they want to notify the other user about replying to their message, kind of a reply vs. quote tweet.

MSC3664 gives the *receiver* control over whether to know if a user replies to their message or not.

I'm not sure if that really matters here and I think could be handled in a follow-up. *Not* doing this would add a dependency on MSC3664, however, to avoid regressing behavior.

Reviewer: Sorry, I did misread the context, but my point still stands: I think we should add a dependency on #3664.

It feels wrong to depend on the sending user's client determine whether or not the receiver's client should notify on a particular reply, and it feels like it would just be a stopgap measure that could be solved by an existing server-side change that gives more control to the user who is actually receiving the notification.

Author: > Sorry, I did misread the context, but my point still stands: I think we should add a dependency on #3664.

I'd like to avoid adding a dependency on an MSC that seems stuck.

> It feels wrong to depend on the sending user's client determine whether or not the receiver's client should notify on a particular reply, and it feels like it would just be a stopgap measure that could be solved by an existing server-side change that gives more control to the user who is actually receiving the notification.

I'm pondering if it makes sense to have *both*, as a thought experiment, if we have clients add to the `mentions` array when replying it could:

* Allow MSC2781 to land without MSC3664. (I'm not arguing this is desirable, just that it is possible.)
* Allow mentioning both `alice` & `bob` when `charlie` replies to `bob`'s reply to `alice`. (Currently this is done via reply-fallbacks IIRC, but that wouldn't be possible after MSC2781, even with MSC3664.)
* Reply *without* mentioning the original sender: "quote-tweet"-ing.

The last two points give control to the *sender*, but for the *receiver* to also have control then MSC3664 could be used:

* Allow a user to be notified of *all* replies to their messages, regardless of if the sender included the user in the `mentions` array.

----

I don't know if this over-complicates things, but I think it is worth at least considering as an alternative or augmentation to MSC3664.

Reviewer 5: As a lay of the chat landscape; Discord has the ability to enable/disable notifying a user when you reply to their message. I personally find this feature quite helpful. For instance, I'll turn off the notification if I'm referencing something someone mentioned earlier, but am directing my message to the immediate conversation (which the person I'm replying to may not be part of).

Author: I'm going to park this conversation -- it feels like a future MSC should work out exactly how replies should work; perhaps I will write an alternative to MSC3664 that depends on this.

Author: Actually -- I'm going to unresolve this thread, I think if we're asking clients to include the `m.mentions` field on all messages we need to deal with how users get notified of replies (otherwise users will not be notified of replies once this MSC lands).

Author: 7de34e42eb0abce53aec842ff7ed11322d406e02 adds a bunch of information about this.

> user that they were mentioned.

## Proposal

Instead of searching a message's `body` for the user's display name or the localpart
of their Matrix ID, it is proposed to use a field specific to mentions,[^3] and
augment the push rules to search for the current user's Matrix ID.

### New event field

A new `mentions` field is added to the event content, it is an array of strings
consistent of Matrix IDs.

To limit the potential for abuse, only the first 10 items in the array should be
considered. It is recommended that homeservers reject locally created events with
more than 10 mentions with an error with a status code of `400` and an errcode of
`M_INVALID_PARAM`.

Clients should add a Matrix ID to this array whenever composing a message which
includes an intentional [user mention](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
(often called a "pill"). Clients may also add them at other times when it is
obvious the user intends to explicitly mention a user.

The `mentions` field is part of the plaintext event body and should be encrypted
into the ciphertext for encrypted events.

### New push rules

> A new push rule condition and a new default push rule will be added:

Author: We need to mention where this push rule is in the list of push rules.

message to 10 or force the user to remove some mentions.

The `mentions` field is part of the cleartext event body and should **not** be
encrypted into the ciphertext for encrypted events. This expoes slightly more
metadata to homeservers, but allows mentions to work properly via push notifications
(which is a requirement for a usable chat application) and should result in bandwidth
and battery savings.

### New push rules

Two new push rule conditions and corresponding default push rule are proposed:

The `is_user_mention` push condition is defined as[^4]:

> This matches messages where the `content.mentions` is an array containing the
> owner’s Matrix ID in the first 10 entries. This condition has no parameters.
> If the `mentions` key is absent or not an array then the rule does not match;
> any array entries which are not strings are ignored, but count against the limit.

The `is_room_mention` push condition is defined as[^4]:

> This matches messages where the `content.mentions` is an array containing the
> string `"@room"` in the first 10 entries. This condition has no parameters.
> If the `mentions` key is absent or not an array then the rule does not match;

> > any array entries which are not strings are ignored, but count against the limit.

Author: I debated making this a `is_mention` condition with a value of some sort, which would be the user's MXID or `@room`. Similar to some other conditions e.g. from `.m.rule.invite_for_me`:

```json
{
    "key": "state_key",
    "kind": "event_match",
    "pattern": "[the user's Matrix ID]"
}
```

Or even more generic, making it a `field_contains` which would take a `key` and `value` (or `pattern`).

Author: Yep, role mentions are a thing I was considering as a nice extension. I wonder if the field would make more sense to use an object and not do magic string matching on `@room`:

```json5
{
    ...,
    "mentions": {
        "user_ids": [...],  // optional
        "room": true,  // optional
        "roles": [...]  // defined in a future MSC
    }
}
```

Reviewer 3: This could maybe be achieved by making it `mention`, and in the same MSC telling that `mentions` is optional and deprecated.

It already is "optional" here due to historic behaviour, but in the future that could be used as a stepping stone for transition to an object?

Author: I don't see why we would define it one way just to immediately deprecate it? If we think we might add to it in the future, then we should just make it an object now.

> # MSC3952: Intentional Mentions

Reviewer 6: Though this would be a pretty significant addition to the MSC, it might be worth considering specifying a Matrix HTML tag also to go with this in the body. For instance:

```
<mx-mention target=0>hello</mx-mention>
```

Where the target is the index of the mxid in the mentions array. This could help deal with spam because mentions not referenced in the body could be ignored (the mentions key would be used as a hint to save battery and the message contents would be checked to confirm whether to send a notification). Furthermore, the tags could be used to indicate which part of the message one wants to being to the attention of the mentioned user. A pill could then also be inserted where the tag starts to allow for direct access to the mentioned user's profile by clicking it, as many chat clients do.

This hybrid approach effectively retains advantages of putting mentions in the body:
- Less opportunity for mention spam
- Easy placement of pills
- Directing attention to parts of messages

And of putting them in a separate key:
- Easier to search hence less resource-hungry
- Less ambiguous

With none of the disadvantages, except maybe a very slightly larger event size.

Author: I spent quite a bit of time considering similar formats, but it was unclear how to handle this for plaintext messages (or other message formats, e.g. if a `format` of `com.example.rff` was used -- what do we do in those cases?

There are additional downside:

* It doesn't allow for hidden mentions, which *can* be useful in some situations.
* It isn't as extensible to event types except `m.room.message`.

I should document this in the alternatives though. 👍 Thank you for the reminder!

Reviewer 6: Fair points, though I find it hard to let go of the ability to place pills in specific places.

What if we make the tag optional?

That would nullify my point about "less opportunity for mention spam" but otherwise has the same advantages and should address your concerns as it would optionally apply to any event which supports matrix HTML tags.

Author: > What if we make the tag optional?

This is essentially how the MSC is currently written, [it recommends](https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/intentional-mentions/proposals/3952-intentional-mentions.md#new-event-field):

> Clients should add a Matrix ID to this array whenever composing a message which includes an intentional [user mention](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions) (often called a "pill").

A pill is HTML of the form of `<a href='https://matrix.to/#/@alice:example.org'>Alice</a>`, i.e. a matrix.to link to the user.

Reviewer 6: I guess it's not exactly the same thing, since it allows for non-mention pills, but that's actually probably a good thing so I can agree with that. Cool!

Author: > I guess it's not exactly the same thing, since it allows for non-mention pills, but that's actually probably a good thing so I can agree with that. Cool!

Yep, I think it makes sense if you're talking *about* someone but don't want to mention them? A common example I've had is that I'm DMing someone and say "Oh hey you should talk to my pal `@phil:example.com` about this!", it would be nice to pillify that into a nice matrix.to link, but they're obviously not in the conversation (so you can't "mention" them).

`content.body` field for the current user's display name
([`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules))
or the localpart of their Matrix ID ([`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)).

There's also a [section about "user and room mentions"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
which defines that messages which mention the current user in the `formatted_body`
of the message should be colored differently:

> If the current user is mentioned in a message (either by a mention as defined
> in this module or by a push rule), the client should show that mention differently
> from other mentions, such as by using a red background color to signify to the
> user that they were mentioned.

## Proposal

Instead of searching a message's `body` for the user's display name or the localpart
of their Matrix ID, it is proposed to use a field specific to mentions,[^3] and
augment the push rules to search for the current user's Matrix ID.

### New event field

A new `mentions` field is added to the event content, it is an array of strings

> consistent of Matrix IDs or the special value: `"@room"`.

Reviewer 5: heh, these were meant to be turned into suggestions before submitting my review.

```suggestion
consisting of Matrix IDs or the special value: `"@room"`.
```

> To limit the potential for abuse, only the first 10 items in the array should be

Reviewer 6: I would argue that this should be increased and pretty significantly too (maybe up to 100 or even 255 entries?). My reasoning is that spammers can be dealt with by means of banning or server ACLs, and between `@room` and just sending multiple events, they also have plenty of ways to sidestep this altogether. Meanwhile, legitimate uses such as mentioning a specific team in one message would be more affected by this, since Matrix has no concept of role mentions or the likes.

In general I can imagine legitimate users hitting this ceiling while I think such protocol-level limits should at least be at a level where we shouldn't expect to find legitimate users.

As you mention further down, making this configurable would address this issue, but too much configurability isn't always a good thing either, though I would prefer that over a limit of 10 people.

Author: 10 was a mostly arbitrary number -- it seemed reasonable for most cases. I wouldn't be opposed to increasing this a bit, but I don't think I've seen a message that wanted to mention 100 people or so though -- that seems like quite a lot?

Reviewer 6: I also would not expect regular messages to have anywhere near 100 mentions, but given that this is a hard limit which would take a long time to increase should it prove to be too restrictive (as the whole client/server ecosystem would need to adopt the new limit) I think too much is better than too little here.

We should also keep in mind bridging and automation when setting the limit. Some platforms we may want to bridge with might allow more/unlimited mentions so this limit might make bridging difficult. Bots might also want to mention a large amount of users (but not all users), say in the case of announcements one can subscribe to, as way more than 10 or even 100 users could subscribe to a particular announcement.

Reviewer 7: I'm actually in agreement here that this limit is probably going to cause lots of problems, an example being a calendar bot for a large team which would easily exceed the 10 user ping. Arguably you could ping the room at that point...but presently in Matrix we've been pinging more than 10 folks legitimately.

Playing devils advocate: I'm not *actually* sure putting a maximum on the ping count is all that useful in the same way we don't limit the number of  `<p>` you can put into a message to fill up a room's timeline. This feels like a thing where if someone tries to spam by pinging lots of users, you just kick the user. (Possibly ending up as a Mjolnir moderation rule, for automatic-ness)

I think hardcoding a maximum ping count in the spec will ultimately break *someones* valid use case, and to me the spec shouldn't be policing user content. It should just facilitate the ability of other services/humans to handle it.

Reviewer 8: As a concrete example, for the Element web weekly we have 120 users in the room and the calendar reminder pings the 17 required attendees.

Reviewer 9: > I think hardcoding a maximum ping count in the spec will ultimately break someones valid use case, and to me the spec shouldn't be policing user content. 

I fundamentally disagree with this because we _do_ need limits to ensure servers can remain quick and responsive. 

That being said, there is an inherent limit anyway due to event size limits. The main question revolves around how expensive the `mentions` field is computationally/memory wise from a server perspective. If it's dead cheap then sure let's have no limit, but if this is expensive / can be abused then not having a limit is just asking for trouble imo.

Reviewer 10: how may can be mentioned in 64k of JSON?

Author: > If it's dead cheap then sure let's have no limit, but if this is expensive / can be abused then not having a limit is just asking for trouble imo.

It is likely fairly cheap; the server (and client) would traverse the list (ignoring non-string entries), and search for the current user's Matrix ID. You could certainly make a pretty long list if this is the only bit of your message size.

----

As I mentioned above, but maybe not strongly enough. **10 was chosen arbitrarily** and I am definitely not opposed to increasing it. I'll look into this further.

Author: I've moved this to a recommendation for moderation tooling for now.

considered. It is recommended that homeservers reject locally created events with
more than 10 mentions with an error with a status code of `400` and an errcode of
`M_INVALID_PARAM`. It is valid to include specific users in addition to the

> `"@room"` value.

Author: This isn't called out, but yes the implication is that it would only ever generate a single notification per message.

Clients should add a Matrix ID to this array whenever composing a message which
includes an intentional [user mention](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
(often called a "pill"). Clients should add the `"@room"` value when making a
room-wide announcement. Clients may also add them at other times when it is
obvious the user intends to explicitly mention a user.

If a user generates a message with more than 10 mentions, the client may wish to
show a warning message to the user; it may silently limit the number sent in the
message to 10 or force the user to remove some mentions.

The `mentions` field is part of the cleartext event body and should **not** be

> encrypted into the ciphertext for encrypted events. This expoes slightly more

Reviewer 5: ```suggestion
encrypted into the ciphertext for encrypted events. This exposes slightly more
```

> metadata to homeservers, but allows mentions to work properly via push notifications

Reviewer 8: Do clients not already have to decrypt events when receiving push notifications, e.g. in order to display the event text in the system notification area?

Author: I tried to make this a bit clearer by pointing to another section of the MSC.

(which is a requirement for a usable chat application) and should result in bandwidth

> and battery savings.

Author: Can this field be edited or shall we declare that it cannot be? What happens if the wrong person is mentioned? Should you be able to fix that?

Author: If we allow this to be edited it:

* Avoids special casing this field.
* Allows people to correct who is mentioned in a message.
  * But this might interact weirdly with MSC3664, which says that edits should not notify? Maybe we shouldn't worry about that at the moment.
* How would you edit this if this is an encrypted event?

Reviewer 6: Though I can't answer the questions you listed, I want to put my vote in to allow edits, for parity with the current way mentions are done.

Author: Right now edits cause notifications to be processed; it doesn't seem to be part of this MSC to change that behavior or define special rules there. I'm going to park this for a future MSC.

every message to avoid the unintentional mentions discussed above.

If users wish to continue to be notified of messages containing their display name
it is recommended that clients create a specific keyword rule for this, e.g. a
`content` rule of the form:

```json
{
  "actions": [
    "notify",
    {
      "set_tweak": "sound",
      "value": "default"
    },
    {
      "set_tweak": "highlight"
    }
  ],
  "pattern": "alice",
  "rule_id": "alice",
  "enabled": true
}
```

After acceptance, it is likely for some disagreement about which push rules are
implemented: legacy clients and homeservers may not yet have deprecated the
`.m.rule.contains_display_name`, `.m.rule.contains_user_name`, and `.m.rule.roomnotif`
push rules, while up-to-date clients and homeservers will support the
`.m.rule.is_user_mention` and `.m.rule.is_room_mention` push rules. It is expected
that both sets of push rules will need to be supported for a period of time, but
at worst case should simply result in the current behavior (documented in the preamble).

## Potential issues

### Abuse potential

This proposal makes it trivial to "hide" mentions since it does not require the

> mentioned Matrix IDs to be part of the displayed text. Depending on the use this

Reviewer 9: This is only "hidden" in the sense that client UIs don't currently display this information. Nothing stops clients from making these visible e.g as a decorator like we have `(edited)` we can easily say `(notified X people)` with a click-through to the list.

Author: Definitely! I think this is even mentioned below (or maybe it was just discussed as part of the team). By "hidden" I meant "not part of the message" content -- I'll try to spell that out a bit more that this is a current UI limitation.

Author: I reworked this section a bit to combine the UI ideas together up-front, then discuss abuse a bit more. I think that flows better. Thanks!

  extremely similar to the proposal in this MSC, but limited to encrypted events.
* [MSC2463](https://github.com/matrix-org/matrix-spec-proposals/pull/2463):
  excludes searching inside a Matrix ID for localparts / display names of other
  users.
* [MSC3517](https://github.com/matrix-org/matrix-spec-proposals/pull/3517):
  searches for Matrix IDs (instead of display name / localpart) and only searches
  specific event types & fields.
* [Custom push rules](https://o.librepush.net/aux/matrix_reitools/pill_mention_rules.html)
  to search for matrix.to links instead of display name / localpart.

  <details>

  The above generates a new push rule to replace `.m.rule.contains_display_name`
  and `.m.rule.contains_user_name`:

  ```json
  {
     "conditions": [
       {
           "kind": "event_match",
           "key": "content.formatted_body",
           "pattern": "*https://matrix.to/#/@alice:example.org*"
       }
     ],
     "actions" : [
       "notify",
       {
         "set_tweak": "sound",
         "value": "default"
       },
       {
         "set_tweak": "highlight"
       }
     ]
  }
  ```

  </details>

> ### Room version for backwards compatibility

Reviewer 11: mentioned over DM, but an option might be to throw all the eggs into one basket here and phrase this MSC as an MSC1767-compatible mixin (similar to MSC3955), which then can be considered "core" for extensible events itself. This would mean that old, non-extensible, room versions don't get a mentions array, but new events and their shiny new format would.

It's unclear to me at this stage if this would be a good idea, but it may be a practical idea to encourage adoption while not having to worry about backwards compatibility.

Author: > it may be a practical idea to encourage adoption while not having to worry about backwards compatibility.

It does solve the backwards compatibility issue nicely, the main downside would be pushing out the timeline for adoption.

Alternative backwards compatibility suggestions included using a new room version,
similar to [MSC3932](https://github.com/matrix-org/matrix-spec-proposals/pull/3932)
for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable
room version before use, which would unnecessarily delay usage.

### Encrypting the `mentions` array

Encrypting the `mentions` array would solve some unintentional mentions, but
will not help with???.

Allowing an encrypted `mentions` array on a per-message basis could allow users
to choose, but would result in inconsistencies and complications:

* What happens if the cleartext `mentions` field does not match the plaintext
  `mentions` field?
* Do push rules get processed on both the cleartext and plaintext message?

It may be argued that clients need to decrypt all messages anyway to handle

> user-specific keywords, the process for doing this is costly (either receiving

Reviewer 8: Clients will also have to decrypt events for rules that filter on the event type, such as the polls-specific push rules in MSC3930. I suspect that live location sharing as defined in MSC3489 will require similar event-specific rules.

Author: Only if they're being pushed every event; in the case where a room is set to mentions-only we could stop pushing those events, see our [above conversation](https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1084916055).

I'm not trying to imply that clients no longer need to decrypt events to process notifications, just that this fixes a class of needing to *push* events.

Reviewer 8: Oh, I see. I hadn't previously understood that you were mainly thinking about rooms that are set to mentions-only. That might be a nice improvement.

Reviewer 8: For location sharing we haven't yet detailed the push rules but for polls MSC3930 proposes underride rules for controlling notifications on poll start / end events but also an _override_ rule for silencing poll votes in general. Do you think that would conflict with your proposal?

Author: No, that should work fine!

  necessary to process the textual `body` for every user's display name and local
  part).

Overall this proposal seems to be neutral or positive in the ability to combat
malicious behavior.

## Future extensions

### Combating abuse

Some ideas for combating abuse came from our discussion and research which are
worth sharing. These ideas are not part of this MSC, but also do not depend on it.

It was recommended that clients could expose *why* an event has caused a notification
and give users inline tools to combat abuse. For example, a client might show a tooltip:

> The sender of the message (`@alice:example.org`) mentioned you in this event.
>
> Block `@alice:example.org` from sending you messages? `[Yes]` `[No]`

Additionally, tooling to for moderators to quickly find messages which mention
many users would be useful. 

A future MSC might wish to explore features for trusted contacts or soft-ignores
to give users more control over who can generate notifications.

### Muted except for mentions push rules

It might be desirable to have a "muted-except-for-mentions" feature for large (encrypted)
rooms. This is particularly useful on iOS where a push notification can be decrypted

> via a background process but *cannot* be suppressed. This means it is not possible

Reviewer 12: I am pretty sure this is possible nowadays: https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering

or the localpart of their Matrix ID ([`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)).

There's also a [section about "user and room mentions"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
which defines that messages which mention the current user in the `formatted_body`
of the message should be colored differently:

> If the current user is mentioned in a message (either by a mention as defined
> in this module or by a push rule), the client should show that mention differently
> from other mentions, such as by using a red background color to signify to the
> user that they were mentioned.

## Proposal

Instead of searching a message's `body` for the user's display name or the localpart
of their Matrix ID, it is proposed to use a property specific to mentions,[^4] and
augment the push rules to search for the current user's Matrix ID.

### New event property

A new `m.mentions` property is added to the event content; it is an object with two
optional properties:

* `user_ids`: an array of strings consisting of Matrix IDs to mention.

> * `room`: a boolean, true indicates an "@room" mention. Any other value or the

Author: I see this as potentially a property under this, see https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/intentional-mentions/proposals/3952-intentional-mentions.md#role-mentions

> * `room`: a boolean, true indicates an "@room" mention. Any other value or the

Reviewer 13: Yeah, my suggestion was more as a means of avoiding having to add a new property in the future, but I think it's more a matter of taste.  While I have a slight preference for my suggestion, either way is fine.

> * `room`: a boolean, true indicates an "@room" mention. Any other value or the

Author: Oh, I've just realized your comment applies *specifically to the `room`* field! That clears it up a bit, to visualize this a bit I think you're saying this would be:

```json
{
  "m.mentions": {
    "user_ids": ["@alice:example.com"],
    "groups": ["m.room"]
  }
}
```

The original text of this MSC (and MSC1796) proposed using a constant (`"@room"`) inside the same array as the user IDs, I liked separating it out for clarity / less magic (and less potential for parser errors). Having an array of other "groups" of people might make sense, but I think it still is clearer to separate them out, e.g.:

```json
{
  "m.mentions": {
    "user_ids": ["@alice:example.com"],
    "room": True,
    "roles": ["m.moderator"],
    "room_ids": ["#parent-space:example.com"]
  }
}
```

I'm struggling a bit without trying to define the behavior there.

> * `room`: a boolean, true indicates an "@room" mention. Any other value or the

Author: @uhoreg I think I would prefer to leave this as-is; do you want me to note this in the alternatives?

  property missing is interpreted as not an "@room" mention.  

It is valid to include both the `user_ids` and `room` properties.

It is recommended that homeservers reject locally created events with an invalid
`m.mentions` property with an error with a status code of `400` and an errcode of
`M_INVALID_PARAM`.

Clients should add a Matrix ID to the `user_ids` array whenever composing a message which
includes an intentional [user mention](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)

> (often called a "pill"). Clients should set the `room` value to `true` when making a

Reviewer 13: Also add the Matrix ID when replying to a message by that user?

Author: There's another thread about this: https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1063689061

I think I'm inclined to say "yes", but others feel that MSC3664 is a better solution. I'd like to see a hybrid approach where the sending client includes it (to signal the the intention that the user should be notified) and then the receiver decides whether it wants to be notified or not via a combination of this MSC + MSC3664. This would allow users to quote-reply, essentially. (I.e. without pinging the initial sender.)

Author: I've unresolved the other thread that has more context (I'm going to resolve this one to not have duplicates!)

Author: Note that 7de34e42eb0abce53aec842ff7ed11322d406e02 adds some info about replies.

room-wide announcement. Clients may also set these values at other times when it is

> obvious the user intends to explicitly mention a user.

Author: An earlier version of the MSC (or maybe the design doc?) discussed that clients could prompt when trying to send a message that will notify X people to double check you really do mean to send that message. I think it would not be a requirement, but does seem like a good idea. I can add a brief note about it.

> obvious the user intends to explicitly mention a user.

Author: FTR I don't think this is really *changing* anything, just making the behavior explicit? I think users already expected pilled usernames to be mentioned.

        }
    ]
}
```

An example event matching both `.m.rule.is_user_mention` (for `@alice:example.org`)
and `.m.rule.is_room_mention` is provided below:

```json
{
    "content": {
        "body": "This is an example mention @alice:example.org",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>This is an example mention</b> <a href='https://matrix.to/#/@alice:example.org'>Alice</a>",
        "msgtype": "m.text",
        "m.mentions": {
            "user_ids": ["@alice:example.org"],
            "room": true
        }
    },
    "event_id": "$143273582443PhrSn:example.org",
    "origin_server_ts": 1432735824653,
    "room_id": "!somewhere:over.the.rainbow",
    "sender": "@example:example.org",
    "type": "m.room.message",
    "unsigned": {
        "age": 1234
    }
}
```

### Backwards compatibility

The [`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules),
[`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules),
and [`.m.rule.roomnotif`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules)
push rules are to be removed.

> To avoid unintentional mentions these rules are modified to only apply when the

Reviewer 13: There seems to be a contradiction here, where you say earlier that those rules "are to be removed", but then you say here that the rules "are modified".  How do we both modify and remove the rules?  Or are you saying that the rules will be removed at a later time (i.e. deprecated)?

Author: > Or are you saying that the rules will be removed at a later time (i.e. deprecated)?

Yes, they're essentially deprecated by this MSC and modified to shut-off if the event content has a `m.mentions` property. I'm hoping this is backwards-compatible enough that once the ecosystem upgrades we can just remove them completely (or do it when we do the extensible events room upgrade?)

Author: Fixed in e346fbb27bf1374f0d58b38ea2d1b1694397c305.

`m.mentions` property is missing; clients should provide the `m.mentions` property on
every message to avoid the unintentional mentions discussed above.

If users wish to continue to be notified of messages containing their display name
it is recommended that clients create a specific keyword rule for this, e.g. a
`content` rule of the form:

```json
{
  "actions": [
    "notify",
    {
      "set_tweak": "sound",
      "value": "default"
    },
    {
      "set_tweak": "highlight"
    }
  ],
  "pattern": "alice",
  "rule_id": "alice",
  "enabled": true
}
```

After acceptance, it is likely for there to be disagreement about which push rules are
implemented: legacy clients and homeservers may not yet have deprecated the
`.m.rule.contains_display_name`, `.m.rule.contains_user_name`, and `.m.rule.roomnotif`
push rules, while up-to-date clients and homeservers will support the
`.m.rule.is_user_mention` and `.m.rule.is_room_mention` push rules. It is expected
that both sets of push rules will need to be supported for a period of time, but
at worst case should simply result in the current behavior (documented in the preamble).

## Potential issues

> 

Reviewer 13: I'm not sure if this is worth mentioning, or where it should be mentioned, but bridges may need to parse messages and add a `m.mentions` when it thinks the sender is mentioning someone.  The bridge may guess incorrectly and ping someone who wasn't supposed to be mentioned (intentionally), but I don't think the situation is any worse than what we currently have.

Author: I thought I included this info in here, but...I didn't. 😄 According to @Half-Shot this MSC has the potential to make things better (as bridges can run into all the same issues discussed in the pre-amble) without really making things worse -- i.e. bridges for protocols without explicit mention information (IRC?) still need to do the mapping between IRC and Matrix ID, now they just have the explicit mapping to put that information into, instead of trying to modify the original message text.

(@Half-Shot or @tulir does that sound accurate?)

Author: I dug up my previous conversation with @Half-Shot about this, it mostly echos [previous thoughts](https://github.com/matrix-org/matrix-spec/issues/353#issuecomment-1055809364) of his:

> Yes, we had the problem where IRC users would ping users on Matrix by their IRC nick, which meant we had to search and replace all messages coming from IRC ->Matrix with your Matrix nick in an effort to make it ping.
>
> Which was a bit fraught and rubbish, because:
>
> 1. You might, if your nickname was something like `will`, turn off keyword matching.
> 2. It broke the contract of sorts we have with bridges not to mutate the content of messages.
> 3. It also meant we had to build Regex's to run over every message, and try to figure out if a message contained all possible nicknames that we had cached in memory.
>
> I don't think this MSC fixes 3, but it does fix 1 and 2.

I'll try to include some of this in the MSC!

Author: Added in 8f2805defcc25865f37e9ddfd94b08553b2bbbe4.

Alternative backwards compatibility suggestions included using a new room version,
similar to [MSC3932](https://github.com/matrix-org/matrix-spec-proposals/pull/3932)
for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable
room version before use, which would unnecessarily delay usage.

### Encrypting the `m.mentions` property

Encrypting the `m.mentions` property would solve some unintentional mentions, but
leaving it unencrypted allows for solving a wider range of issues.

Allowing an encrypted `m.mentions` property on a per-message basis could allow users
to choose, but would result in inconsistencies and complications:

* What happens if the cleartext `m.mentions` property does not match the plaintext
  `m.mentions` property?
* Do push rules get processed on both the cleartext and plaintext message?

It may be argued that clients need to decrypt all messages anyway to handle
user-specific keywords. This is a costly process in terms of bandwidth, CPU, and
battery since the client must either receive every message via push notifications

> or backpaginate every room fully, in case of a gappy sync.

Reviewer 14: According to @ara4n, we *do* expect clients to decrypt all messages nowadays (and for all encrypted messages to cause pushes so that they can do so).  So it seems like the argument against encryption largely goes away?

It is asserted that the use-cases for custom keywords and mentions are sufficiently
different that having different solutions, and different urgencies to receiving
those notifications, is reasonable. In particular, custom keywords is more of a
power-user feature while user mentions working properly (and promptly) is a table-stakes
feature for a messaging application.

## Security considerations

Including the mentioned users in cleartext is a small metadata leak, but we consider that the tradeoff
of mentions working properly and the savings in bandwidth and battery makes it worthwhile.

The additional metadata leaked is minor since the homeserver already knows the
room members, how often users are sending messages, when users are receiving messages
(via read receipts, read markers, and sync), etc. Exposing explicit mentions does not

> significantly increase metadata leakage.

Reviewer 15: That said, I also believe this *not* to be an appropriate design trade-off for several reasons.

First, Matrix already leaks far too much (meta)data which should be encrypted and this is a move in the wrong direction. It's well known that a lot meaning can be gleaned contextually from even simple things like timestamps and message senders, let alone the more juicy stuff like message types, reactions and who the message mentions. If Matrix is to be taken seriously in circles requiring strong security, we should be working on reducing this surface area rather than increasing it.

Second, there is clearly a fundamental tension between encryption and server-side processing. Every time a feature is implemented on the server, it necessarily means we must be leaking some data to it. But there are already existing important Matrix features which require complete client-side timeline access in order to work at all -- notable examples include encrypted room search and keywords (as mentioned by the MSC too). Unless we plan on dropping support for these features, we will already have all required information to implement properly encrypted mentions. This should at least give us some pause.

Finally, there is a game theoretical problem here. Some people will want to encrypt mentions, and there is nothing preventing them from implementing this. I am predicting this will happen anyway, so it would be nice if this MSC was deliberately crafted in such a way to allow for and encompass this usage too.

Ultimately, I believe encrypted room search is a core feature that most users will want and so support for it will grow with time, at which point cleartext mentions in encrypted rooms will be an unnecessary sore thumb. At a minimum, I think this MSC shouldn't be making the call to place explicit mentions in cleartext but allow for both cleartext and plaintext placement.

> significantly increase metadata leakage.

Reviewer 16: on reflection, I agree with @dkasak completely on this - @clokep, i'm really sorry for not having bandwidth to properly review this and feed back sooner.

The thing that clinches it for me is that we already are committed to downloading all contents in E2EE rooms in order to calculate unread state (given the server doesn't know if a msg is visible until it's decrypted), notifying for keyword mentions, updating full-text indexes for E2EE content, etc.  For instance, Sliding Sync is designed with this assumption in mind: https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/sync-v3/proposals/3575-sync.md#e2ee-handling.

> significantly increase metadata leakage.

Reviewer 13: I agree with @dkasak here.  Making the mentions unencrypted makes me uneasy.  It's not clear that the benefits outweigh the data leak risk, and I'd prefer to err on the side of security here.

@mscbot concern Security of unencrypted mentions

> significantly increase metadata leakage.

Author: Thanks for leaving your thoughts, I had originally written this MSC without any special encryption behavior for this property (i.e. it would be encrypted), but
updated it [based on a conversation](https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1063691992) which unearthed MSC1796. From my reading of that it seemed the logic in MSC1796 was still sound, *and* it would allow for additional features (namely "muted-except-for-mentions" push rules). The previous thread did discuss a bit about making it configurable whether to encrypt or not, but that seemed overly complicated.

I suppose MSC1796 should have been closed when matrix-org/synapse#1131 was merged into Synapse, i.e. once we started pushing all messages to mobile devices. (Note that this was before the MSC process so I couldn't really figure out how much we wanted this was vs. it being a hack to make things work.)

> The thing that clinches it for me is that we already are committed to downloading all contents in E2EE rooms in order to calculate unread state (given the server doesn't know if a msg is visible until it's decrypted), notifying for keyword mentions, updating full-text indexes for E2EE content, etc.

I will *softly* push back on this though: mobile clients do get all encrypted messages pushed to them so they do properly calculate unread state; although they do not always do proper keyword matching *or* full-text indexing of this content it seems they *could*.

Most[^1] web clients, however, *do not* get all encrypted messages and only do a partial backfill on gappy syncs (https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/sync-v3/proposals/3575-sync.md#e2ee-handling describes this briefly); thus the unread state (and keyword notifications) are an estimate.

The full-text index on Element Desktop (I'm unsure if any other clients have this feature) *does* fully backfill, but does not share this information with the notification processing (and likely has different assumptions, e.g. it backfills slowly in the background).

----

tl;dr: I'll revert the changes to this MSC which discuss keeping the `m.mentions` field in cleartext. I think the rest of the implementation is still sound, but pushes a bit more work onto clients.

[^1]: The only web client I know of which gets push notifications is [Hydrogen](https://github.com/vector-im/hydrogen-web/blob/master/doc/implementation%20planning/PUSH.md), but it [doesn't seem to support decrypting pushed messages](https://github.com/vector-im/hydrogen-web/blob/13428bd03c7ec3821352ab13eb631fb0bbe23e94/src/platform/web/sw.js#L297-L303).

> significantly increase metadata leakage.

Reviewer 2: Agreed on reverting the controversial parts to get the MSC through, but I still do want to advocate for servers being able to discern when to push notifications, especially in the iOS case.

In iOS, in order to display an encrypted notification to the user, we need to use a UNNotificationServiceExtension (https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension), which allows us to run some code to decrypt a notification to display the cleartext content to a user. A drawback to this approach is that iOS insists that every notification that is pushed is shown to the user, and the UNNotificationServiceExtension (aka NSE) is only allowed to transform content, not supress. This means that unless I want a push notification on my home screen for every encrypted message in a group chat I'm in, I need to mute the room using a push rule to mute all notifications to that room. This would override https://github.com/matrix-org/synapse/pull/1131, which brings us to the situation where we're at today, where mentions in muted rooms get missed.

Having the the mention information in the cleartext would allow us to implement on the server an APNS push for mentions but not for other encrypted messages.

This is probably a discussion for a future MSC, and again, I think it makes sense to remove putting the mention information in the cleartext to get this through, but in the future I'd like to allow the clients that our users use to specify which metadata they're ok leaking to opt into better UX, depending on their usecase.

> significantly increase metadata leakage.

Author: Thanks for more of the technical information.

I've updated the text in f0a1f6ad184788814c45d58370248b8052142171 to give no special handling of this field. A future MSC may wish to do so.

> significantly increase metadata leakage.

Reviewer 15: Provided it's configurable, I am considerably less closed to it. But that would also make it significantly less reliable, since unless the user enables it for their mentions, push notifications would still not work on iOS under the described implementation.

The real problem seems to be that iOS is such a user and encryption hostile platform.

> significantly increase metadata leakage.

Reviewer 12: I am pretty sure you can hide notifications on iOS nowadays: https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering

Does anyone have other information?

> significantly increase metadata leakage.

Reviewer 11: (marking as unresolved given this area is a *very* active discussion)

> significantly increase metadata leakage.

Reviewer 2: > I am pretty sure you can hide notifications on iOS nowadays: https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_usernotifications_filtering

Fair point! Looks like the request is specifically for end-to-end encrypted apps, and should be fairly straightforward for iOS Matrix clients to request and receive.

This means that room mutes should be applied with push rules at a lower priority than https://github.com/matrix-org/synapse/pull/1131, and then it would be the NSE's responsibility to decrypt and filter out any notifications that are for muted rooms but don't have mentions once the cleartext is known.

> significantly increase metadata leakage.

Reviewer: Even if it's possible, evaluating mentions client-side is still quite expensive, as most encrypted events will not be mentions. We would be expending a lot of energy to decrypt many messages which won't ever notify which would drain battery.

> significantly increase metadata leakage.

Reviewer 17: I can see the wisdom in leaking the least amount of information possible in encrypted rooms, after all, that is the reason there is encryption enabled in the first place.  There should be [warnings in the clients stating that additional bandwidth/energy will be consumed as MSC 1796 recommends under the custom keywords section](https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/msc1796/proposals/1796-e2e-notifications.md#better-handling-of-custom-keyword-notifications), although this should be amended to include Mention notifications as well if they do end up encrypted.  Unfortunately at this time it looks like there will have to be UX and usability sacrifices to keep E2EE actually secure (leaking less message contents/metadata), so there should be an effort to state the consequences of enabling encryption on a group room before letting the user flick the switch (especially since it is irreversible.)

> significantly increase metadata leakage.

Author: My overall take on this is that: MSC3952 as it is written today fixes *several* issues with how users get mentioned on Matrix. It should hopefully:

* Reduce noise from unwanted notifications (mostly due to technical gremlins, but also by making it possible for clients to surface *exactly* who will be mentioned by a user).
* Simplifies parsing of messages to see whether it should be a mention.
* Make it easier for less technical users to understand how mentions work (i.e. we won't need to use leetspeak or weird display names to get around bugs).

----

It does *not* (and never attempted **as a primary goal**) to solve any of the following (which this thread seems to have taken a turn towards):

* Attempt to dramatically reduce battery or data usage of mobile devices.
* Attempt to help fill notification holes in gappy-sync situations for non-push clients (i.e. web clients).
* Enable mute-except-for-mentions push rules.

These are all left as "status quo" although I think [some of the ideas in this MSC](https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/intentional-mentions/proposals/3952-intentional-mentions.md#future-extensions) could be leveraged to improve these items I think it is best left to a future MSC which directly targets those problems.

> significantly increase metadata leakage.

Author: Note that I consider this thread resolved, but I'm going to leave it open for now in case folks want to chime in.

> significantly increase metadata leakage.

Reviewer 11: Having been involved in the discussion around the SCT and an attempt to summarize it out of band, I would agree that @clokep's description of the scope is ideal for this MSC. The review process for an MSC is meant to ensure we aren't making things *worse* rather than try to fix everything in one shot - it's deliberately an iterative process, and this MSC's existing scope fits nicely into such an iteration.

Obviously, other SCT members are welcome to disagree on this conclusion, however in re-reading the SCT's internal discussion on this MSC I don't get the sense that the drawbacks listed (having to spider, battery usage, etc) are categorically worse with this change in the eyes of the SCT. 

Plus, this has product team approval from the folks over at Element, indicating that the overall usability of the feature with known technical defects is sufficiently good enough to ship in a product. Element doesn't get to make the total vote, though: there is an open question around whether the current implementation is suitable for this MSC to make forward progress - please take concerns relating to this to [this thread](https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1120888906) on the MSC.

> significantly increase metadata leakage.

Reviewer 18: @dkasak @uhoreg @turt2live  In order to reduce the bandwidth/CPU/time/battery consumption client side, don't you think acceptable to add a unencrypted boolean value to the encrypted event to let the server know there are some mentions inside or not? 
The server will then be able to notify the client only for the events with mentions (without knowing who are actually mentioned). This will help us to restore on mobile the support of the "mention only" notification mode, by limiting the number of notifications received on mobile.

If the end user enabled notifications on some keywords, the server will have to send all events in encrypted rooms, but custom keywords are fairly rare.

> significantly increase metadata leakage.

Reviewer 11: The battery life considerations appear to have already been resolved? As far as the SCT is concerned though, battery life is not a strong enough reason to support this information being unencrypted (unless of course there's no workaround and it ends up using a significant amount of battery, in which case we can reconsider).

> significantly increase metadata leakage.

Reviewer 18: FYI I discussed with @clokep, you may ignore my remark for this MSC. We should handle it in another MSC

[future extensions](#muted-except-for-mentions-push-rules) section). Additionally,
it may be useful for the homeserver to filter or prioritize rooms based on whether
the user has been mentioned in them, e.g. for an extension to
[MSC3575 (sliding sync)](https://github.com/matrix-org/matrix-spec-proposals/pull/3575).

### New push rules

Two new push rule conditions and corresponding default push rule are proposed:

The `is_user_mention` push condition is defined as[^5]:

> This matches messages where the `m.mentions` field of the `content` contains a
> field `user_ids` which is an array containing the owner’s Matrix ID. This
> condition has no parameters. If the `m.mentions` key is absent; not an object;
> does not contain a `user_ids` field; or contains a non-array `user_ids` field
> then the rule does not match. Any non-string array entries are ignored and
> duplicate entries are ignored.

The `is_room_mention` push condition is defined as[^5]:

> This matches messages where the `m.mentions` field of the `content` contains
> a `room` field set to `true`. This condition has no parameters. If the
> `m.mentions` key is absent; not an object; does not contain a `room` field;

> > or contains a non-boolean `room` field then the rule does not match.

Author: I failed to call out in here that the user sending the `@room` must still abide by the power-levels event. There's no intention to change that in this MSC.

Author: Oh, never mind -- I did include the check against `sender_notification_permission`, it is just a different condition! 👍

|---------------------|-------------------|--------------------------------------|
| Event field         | `m.mentions`      | `org.matrix.msc3952.mentions`        |
| Push rule ID        | `.m.rule.*`       | `.org.matrix.msc3952.*`              |
| Push condition kind | `is_user_mention` | `org.matrix.msc3952.is_user_mention` |
| Push condition kind | `is_room_mention` | `org.matrix.msc3952.is_room_mention` |

If a client sees this rule available it can choose to apply the custom logic discussed
in the [backwards compatibility](#backwards-compatibility) section.

If a client sees the *stable* identifiers available, they should apply the new
logic and start creating events with the `m.mentions` field.

## Dependencies

N/A

[^1]: This MSC does not attempt to solve this problem, but [MSC2781](https://github.com/matrix-org/matrix-spec-proposals/pull/2781)
proposes removing reply fallbacks which would solve it. Although as noted in
[MSC3676](https://github.com/matrix-org/matrix-spec-proposals/pull/3676) this
may require landing [MSC3664](https://github.com/matrix-org/matrix-doc/pull/3664)
in order to receive notifications of replies.

> [^2]: Note that this MSC does not attempt to solve the problem of issues generating

Reviewer 2: ```suggestion
[^2]: Note that this MSC does not attempt to solve the problem of edits generating
```

  then the push rule matching usernames (`.m.rule.contains_user_name`) matches
  too often (e.g. Travis CI matching if your Matrix ID is `@travis:example.com`).
* If the [localpart or display name of your Matrix ID matches the hostname](https://github.com/matrix-org/matrix-spec-proposals/issues/2735)
  (e.g. `@example:example.com` receives notifications whenever `@foo:example.com`
  is replied to).

As a sender you do not know if including the user's display name or Matrix ID would
even be interpreted as a mention (see [issue 353](https://github.com/matrix-org/matrix-spec/issues/353)).
This results in some unexpected behavior and bugs:

* Matrix users use "leetspeak" when sending messages to avoid mentions (e.g.
  referring to M4tthew instead of Matthew).
* It is impossible to ping one out of multiple people with the same localpart
  (or display name).
* Since the relation between `body` and `formatted_body` is ill-defined and
  ["pills" are converted to display names](https://github.com/matrix-org/matrix-spec/issues/714),
  this can result in missed messages. [^3]

> There is also some other related bugs:

Reviewer 14: ```suggestion
There are also some other related bugs:
```

Mentions are powered by two of the default push rules that search an event's
`content.body` field for the current user's display name
([`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules))
or the localpart of their Matrix ID ([`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)).

There's also a [section about "user and room mentions"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
which defines that messages which mention the current user in the `formatted_body`
of the message should be colored differently:

> If the current user is mentioned in a message (either by a mention as defined
> in this module or by a push rule), the client should show that mention differently
> from other mentions, such as by using a red background color to signify to the
> user that they were mentioned.

## Proposal

Instead of searching a message's `body` for the user's display name or the localpart
of their Matrix ID, it is proposed to use a field specific to mentions,[^4] and
augment the push rules to search for the current user's Matrix ID.

### New event field

> A new `m.mentions` field is added to the event content, it is an object with two

Reviewer 14: comma splice 
```suggestion
A new `m.mentions` field is added to the event content; it is an object with two
```

(often called a "pill"). Clients should set the `room` value to `true` when making a
room-wide announcement. Clients may also set these values at other times when it is
obvious the user intends to explicitly mention a user.

The `m.mentions` field is part of the cleartext event body and should **not** be
encrypted into the ciphertext for encrypted events. This exposes slightly more
metadata to homeservers, but allows mentions to work properly. It allows the
server to properly handle push notifications (which is a requirement for a usable
chat application) and could result in bandwidth and battery savings (see the
[future extensions](#muted-except-for-mentions-push-rules) section). Additionally,
it may be useful for the homeserver to filter or prioritize rooms based on whether
the user has been mentioned in them, e.g. for an extension to
[MSC3575 (sliding sync)](https://github.com/matrix-org/matrix-spec-proposals/pull/3575).

### New push rules

Two new push rule conditions and corresponding default push rule are proposed:

The `is_user_mention` push condition is defined as[^5]:

> > This matches messages where the `m.mentions` field of the `content` contains a

Reviewer 14: s/field/property/, please, per https://github.com/matrix-org/matrix-spec/blob/main/meta/documentation_style.rst#general.

> `m.mentions` key is absent; not an object; does not contain a `room` field;
> or contains a non-boolean `room` field then the rule does not match.

The proposed `.m.rule.is_user_mention` override push rule would appear directly
before the `.m.rule.contains_display_name` push rule:

```json
{
    "rule_id": ".m.rule.is_user_mention",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "is_user_mention"
        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "sound",
            "value": "default"
        },
        {
            "set_tweak": "highlight"
        }
    ]
}
```

The proposed `.m.rule.is_room_mention` override push rule would appear directly
before the `.m.rule.roomnotif` push rule:

```json
{
    "rule_id": ".m.rule.is_room_mention",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "is_room_mention"
        },
        {
            "kind": "sender_notification_permission",
            "key": "room"
        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "sound",
            "value": "default"

>         },

Reviewer 14: I note that the existing `.m.rule.roomnotif` rule is silent by default (it lacks a `set_tweak: sound` action). Is this a deliberate change?

By leaving the mentions array unencrypted it allows for a "muted-except-for-mentions"
push rule. This is particularly useful for large (encrypted) rooms where you would
like to receive push notifications for mentions *only*. It is imperative for
the homeserver to be able to send a push in this case since some mobile platforms,
e.g. iOS, do not sync regularly in the background.

### Pillifying `@room`

Some clients attempt to create a "pill" out of `@room` mentions, but this is not
a requirement of the Matrix specification. The current [user and rooms mentions](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
section could be expanded for this situation.

### Extensible events

The `m.mentions` field can be considered a "mixin" as part of extensible events
([MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767)) with no needed
changes.

### Role mentions

It is possible to add additional fields to the `m.mentions` object, e.g. a foreseeable
usecase would be a `roles` field which could include values such as `admins` or
`mods`. Defining this in detail is left to a future MSC.

## Alternatives

### Prior proposals

> There's a few prior proposals which tackle subsets of the above problem:

Reviewer 14: ```suggestion
There are a few prior proposals which tackle subsets of the above problem:
```

       "notify",
       {
         "set_tweak": "sound",
         "value": "default"
       },
       {
         "set_tweak": "highlight"
       }
     ]
  }
  ```

  </details>

The last two proposals use a similar idea of attempting to find "pills" in the
`formatted_body`, this has some downsides though:

* It doesn't allow for hidden mentions, which can be useful in some situations.
* It does not work for event types other than `m.room.message`.

It also adds significant implementation complexity since the HTML messages must
now be parsed for notifications. This is expensive and introduces potential
security issues.

### Room version for backwards compatibility

Alternative backwards compatibility suggestions included using a new room version,
similar to [MSC3932](https://github.com/matrix-org/matrix-spec-proposals/pull/3932)
for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable
room version before use, which would unnecessarily delay usage.

### Encrypting the `m.mentions` field

Encrypting the `m.mentions` field would solve some unintentional mentions, but

> will not help with???.

Reviewer 14: looks like this needs finishing up

Allowing an encrypted `m.mentions` field on a per-message basis could allow users
to choose, but would result in inconsistencies and complications:

* What happens if the cleartext `m.mentions` field does not match the plaintext
  `m.mentions` field?
* Do push rules get processed on both the cleartext and plaintext message?

It may be argued that clients need to decrypt all messages anyway to handle
user-specific keywords, the process for doing this is costly (either receiving
every message via push notifications or backpaginating every room fully, in case

> of a gappy sync) in terms of bandwidth, CPU, and battery.

Reviewer 14: I'm struggling to grok this run-on sentence. Rephrase, with fewer commas?

Author: Hopefully 425d1d61a8c494ee2bb9e9f47336f497ba9ee8a2 clarifies this by splitting it into two sentences.

It is asserted that the use-cases for custom keywords and mentions are sufficiently
different that having different solutions, and different urgencies to receiving
those notifications, is reasonable. In particular, custom keywords is more of a
power-user feature while user mentions working properly (and promptly) is a table-stakes
feature for a messaging application.

### For the `is_room_mention` condition

The `is_room_mention` could be replaced with the `exact_event_match` condition
from [MSC3758](https://github.com/matrix-org/matrix-spec-proposals/pull/3758)
(or the changes to `event_match` in [MSC3862](https://github.com/matrix-org/matrix-spec-proposals/pull/3862)),
although either of those would potentially run into ambiguity in the `key`,
as described in [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873):

```json
{
  "kind": "exact_event_match",

>   "key": "m.mentions.room",

Reviewer 14: ```suggestion
  "key": "content.m.mentions.room",
```

  "value": true
}

> ```

Reviewer 14: I'd be quite in favour of this, to reduce the number of push conditions we end up with, and because I think MSC3758 deserves more love than it's had so far.

I realise it means you end up blocked behind MSC3758 (and possibly MSC3873), though, so I'll forgive you if you need to forge ahead.

Author: I agree that MSC3758 is a better solution here, it had not received much interest AFAICS. I'll take another look at it.

Author: This was done in ced04e19730283b6c56a8ef903d20634cde63363.

## Security considerations

> Including the mentioned users in cleartext is a small metadata leak, but the tradeoff
> of mentions working properly and the savings in bandwidth and battery makes it worthwhile.

Reviewer 14: debatable.

```suggestion
Including the mentioned users in cleartext is a small metadata leak, but we consider that the tradeoff
of mentions working properly and the savings in bandwidth and battery makes it worthwhile.
```

The additional metadata leaked is minor since the homeserver already knows the
room members, how often users are sending messages, when users are receiving messages
(via read receipts, read markers, and sync), etc. Exposing explicit mentions does not
significantly increase metadata leakage.

## Unstable prefix

During development the following mapping will be used:

| What                | Stable            | Unstable                             |
|---------------------|-------------------|--------------------------------------|
| Event field         | `m.mentions`      | `org.matrix.msc3952.mentions`        |
| Push rule ID        | `.m.rule.*`       | `.org.matrix.msc3952.*`              |
| Push condition kind | `is_user_mention` | `org.matrix.msc3952.is_user_mention` |
| Push condition kind | `is_room_mention` | `org.matrix.msc3952.is_room_mention` |

If a client sees this rule available it can choose to apply the custom logic discussed
in the [backwards compatibility](#backwards-compatibility) section.

If a client sees the *stable* identifiers available, they should apply the new

> logic and start creating events with the `m.mentions` field.

Reviewer 14: stable identifiers for the push rules?

It feels a little bit magical to me that *sending* of `m.mentions` properties depends on which push rules are seen. Why do we need to gate the transmission of `m.mentions` at all - can't we start immediately?

Why do we need to gate the `m.mentions`

Author: I don't think it was my intention to necessarily gate adding these to events, will reword.

Author: Honestly I have no idea what this sentence is trying to say -- it seems like this was already covered in the backwards compatibility section above about when to apply these push rules.

Author: (I removed it in e346fbb27bf1374f0d58b38ea2d1b1694397c305.)

## Dependencies

N/A

[^1]: This MSC does not attempt to solve this problem, but [MSC2781](https://github.com/matrix-org/matrix-spec-proposals/pull/2781)
proposes removing reply fallbacks which would solve it. Although as noted in
[MSC3676](https://github.com/matrix-org/matrix-spec-proposals/pull/3676) this
may require landing [MSC3664](https://github.com/matrix-org/matrix-doc/pull/3664)
in order to receive notifications of replies.

[^2]: Note that this MSC does not attempt to solve the problem of edits generating
spurious notifications.

[^3]: It is [defined as](https://spec.matrix.org/v1.5/client-server-api/#mroommessage-msgtypes)
the the "plain text version of the HTML [`formatted_body`] should be provided in the `body`",
but there is no particular algorithm to convert from HTML to plain text *except*
when converting mentions, where the
[plain text version uses the link anchor, not the link](https://spec.matrix.org/v1.5/client-server-api/#client-behaviour-26).

[^4]: As proposed in [issue 353](https://github.com/matrix-org/matrix-spec/issues/353).

[^5]: A new push condition is necessary since none of the current push conditions

> (e.g. `event_match`) can process arrays.

Author: I had the same thought at https://github.com/matrix-org/matrix-spec-proposals/pull/3952#discussion_r1067403177; it sounds like @anoadragon453 also slightly favors a generic rule.

> (e.g. `event_match`) can process arrays.

Author: This seems to be MSC3887.

> (e.g. `event_match`) can process arrays.

Author: I wrote MSC3966 for this, will update this MSC with the requirement soon.

This results in some unexpected behavior and bugs:

* Matrix users use "leetspeak" when sending messages to avoid mentions (e.g.
  referring to M4tthew instead of Matthew).
* It is impossible to ping one out of multiple people with the same localpart
  (or display name).
* Since the relation between `body` and `formatted_body` is ill-defined and
  ["pills" are converted to display names](https://github.com/matrix-org/matrix-spec/issues/714),
  this can result in missed messages. [^1]

There are also some other related bugs:

* Matrix users will append emoji or other unique text in their display names to
  avoid unintentional pings.
* Bridging mentions is suboptimal since they [use display names](https://github.com/matrix-org/matrix-spec/issues/353#issuecomment-1055809364)
  as a workaround, e.g.:
  * It broke the contract that bridges will not mutate the content of messages.
  * For some protocols, bridges need try to figure out if every message contains
    any of the possible nicknames of room members.
* If a user changes their display name in a room, they might not be mentioned
  unless the historical display name is used while processing push rules.

>   (TODO I think there's an issue about this?)

Author: https://github.com/matrix-org/matrix-spec/issues/353#issuecomment-1055809372 -- I'll update the reference to it.

## Background

Mentions are powered by two of the default push rules that search an event's
`content.body` property for the current user's display name
([`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules))
or the localpart of their Matrix ID ([`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)).

There's also a [section about "user and room mentions"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
which defines that messages which mention the current user in the `formatted_body`
of the message should be colored differently:

> If the current user is mentioned in a message (either by a mention as defined
> in this module or by a push rule), the client should show that mention differently
> from other mentions, such as by using a red background color to signify to the
> user that they were mentioned.

## Proposal

> Instead of searching a message's `body` for the user's display name or the localpart

Author: Done in 3461ef8f83ac3a5baa433d3e2c954633c06e47b2.

          "event_id": "$initial_event"
      }
    }
  },
  // other fields as required by events
}
```

This signals that it is the *intention* of the sender to mention all of those people,
clients may wish to allow users to modify the list of people to include, e.g. to
"quote reply" as opposed to replying directly.

If a user wishes to be notified of *all replies* to their messages, other solutions
should be investigated, such as [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).
This would give more equal power to both senders and receivers of events.

### Impact on edits

Similarly to [replies](#impact-on-replies), users are notified of message edits
via the `.m.rule.contains_display_name` or the `.m.rule.contains_user_name` push
rule matching the [fallback content](https://spec.matrix.org/v1.6/client-server-api/#event-replacements).
Generally this is undesirable and users do not need to be notified for the same
message multiple times (e.g. if a user is fixing a typo). It is recommended that
clients include an empty `m.mentions` property when editing an event, *unless*
the edit is significant or if additional users are mentioned.

> The full list of mentioned users should be included in the `m.new_content` property.

Author: Yes, although note that there's no way to "cancel" the notification from the initial message. Not sure if that's worth calling out?

> The full list of mentioned users should be included in the `m.new_content` property.

Author: f5dee4986f9760e098ded131390fc9bf5df42504 adds this, it does make me wonder a bit if you add and remove and add and remove the same person. Maybe this is such an edge-case we shouldn't matter though.

          // Include all mentioned users.
          "@alice:example.org",
          "@bob:example.org"
        ]
      },
    },
    "m.relates_to": {
      "rel_type": "m.replace",
      "event_id": "$initial_event"
    }
  },
  // other fields as required by events
}
```

This should limit duplicate, unnecssary notifications for users. If a user wishes
to receive notifications for edits of events they were mentioned in then they
could setup a push rule for the `content.m\\.new_content.m\\.mentions` property
or potentially leverage [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).

### Impact on bridging

This should strengthen the bridging contract as it will allow them to stop
mutating the content of messages during the bridging process (by directly
listing any mentioned users in the `m.mentions` property).

This should reduce the impedance mismatch when bridging protocols which similarly
list out users to mention (the bridge should simply be able to map from the
remote user ID to the bridged user ID)[^4], but will not help in cases where the
mention information is embedded into the text content (as today's Matrix messages
are).[^5]

> ## Potential issues

Reviewer 11: @matrix-org/spec-core-team - please review this thread.

> ## Potential issues

Author: Twitter provides [an `entities` field](https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/tweet) which contains the location into the text content of the Tweet for various things (e.g. links, but also mentions). This model doesn't fit well into Matrix which can have multiple markups, however.

If thi sis just about highlighting *your own* mentions it should be possible to find the corresponding pill (or keyword) and highlight it somehow, but there might not be one. I don't know what you do in this case.

Overall, I suspect this doesn't make it much more difficult to display the mentioning text in a special way since the current mentioning system generally uses a different field than what is displayed to the user (`body` vs. `formatted_body`). Could definitely use some verification, however!

> ## Potential issues

Author: I think with a little tweaking I can get an Element Web preview up that will work for testing this (it should work completely fine in encrypted rooms, even without server support). I *have* updated the matrix-js-sdk & matrix-react-sdk PRs to match the MSC though, so it seems quite doable implementation wise.

Will get back when there's a thing to test.

> ## Potential issues

Reviewer 11: It would be around highlighting your own mentions, with the concern being that other clients will slowly (over years) start to be less and less specific in the details they include in the markup for mentions, thus making it more difficult for receiving clients to find the mention. Currently the approach is to look for the pill, but pills are optional and clients could just as easily rely on localpart. For example, in "travis: have you tried Travis CI?" it's easy for humans to see the mention but the parsing code suddenly becomes more complex to handle natural language.

> ## Potential issues

Author: > Note that I am actively working on that PR so I'll try not to break it, but it might explode. 😄

I think I've gotten most of the bugs out of this PR now (and wrote some tests for it), so it should now be reasonable to test against.

> ## Potential issues

Reviewer 11: We can probably course correct this as needed - resolving for now.

> ## Potential issues

Reviewer 11: (leaving this thread open, but moving ahead with FCP anyways - SCT members who prefer to have more implementation first should `@mscbot concern More implementation`)

mentioned Matrix IDs to be part of the displayed text. This is only a limitation
for current clients: mentions could be exposed in the user interface directly.
For example, a de-emphasized "notified" list could be shown on messages, similar
to CCing users on an e-mail.

Although not including mentions in the displayed text could be used as an abuse
vector, it does not enable additional malicious behavior than what is possible
today. From discussions and research while writing this MSC there are moderation
benefits to using a separate property for mentions:

* The number of mentions is trivially limited by moderation tooling, e.g. it may
  be appropriate for a community room to only allow 10 mentions. Events not abiding
  by this could be rejected automatically (or users could be banned automatically).
* Various forms of "mention bombing" are no longer possible.
* It is simpler to collect metrics on how mentions are being used (it is no longer
  necessary to process the textual `body` for every user's display name and local
  part).

Overall this proposal seems to be neutral or positive in the ability to combat
malicious behavior.

## Future extensions

> ### Combating abuse

Author: I think a previous version of this text including information about notifying when mentioning many people. Seems to have gotten removed. Note that this is likely something we *could do today*, this MSC might make it a bit easier though.

### Pillifying `@room`

Some clients attempt to create a "pill" out of `@room` mentions, but this is not
a requirement of the Matrix specification. The current [user and rooms mentions](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
section could be expanded for this situation.

### Extensible events

The `m.mentions` property can be considered a "mixin" as part of extensible events
([MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767)) with no needed
changes.

### Role mentions

It is possible to add additional properties to the `m.mentions` object, e.g. a foreseeable
usecase would be a `roles` property which could include values such as `admins` or
`mods`. Defining this in detail is left to a future MSC.

## Alternatives

### Prior proposals

There are a few prior proposals which tackle subsets of the above problem:

* [MSC1796](https://github.com/matrix-org/matrix-spec-proposals/pull/1796):
  similar to the proposal in this MSC, but limited to encrypted events (and kept
  in cleartext).
* [MSC2463](https://github.com/matrix-org/matrix-spec-proposals/pull/2463):
  excludes searching inside a Matrix ID for localparts / display names of other

>   users.

Reviewer 11: For tracking: these two MSCs have been added to the SCT workflow as a P1 conflict resolution - members should consider the merits of these two approaches in contrast to this MSC when doing a complete review.

MSC3517 is *not* part of the conflict resolution process because it was closed as obsolete, implying it was adopted/melded into this MSC or similar, thus does not need explicit consideration. Members are welcome to take a look in any case though.

Similarly, https://o.librepush.net/aux/matrix_reitools/pill_mention_rules.html has *not* joined the conflict resolution process as it's not currently recorded as an MSC or proto-MSC. If this is unexpected/untrue, please raise it here.

Reviewer 11: @matrix-org/spec-core-team - please review this thread.

Author: Just wanted to confirm that all sounds true. MSC3517 was closed by the author (deferring to this MSC). the `pill_mention_rules` was never written up as an MSC (and the author has no intention of it, I believe). It is discussed in the alternatives section a bit though.

Reviewer 11: Due to lack of comment, I'm assuming we intend to go ahead with this MSC over the alternatives. SCT members should raise concerns if they disagree.

(leaving this thread open for the moment)

Reviewer 5: IMO, this proposal solves the most edge cases, and introduces more beneficial primitives (`m.mentions)`, over the solutions proposed in MSC3517 and https://o.librepush.net/aux/matrix_reitools/pill_mention_rules.html.

       {
           "kind": "event_match",
           "key": "content.formatted_body",
           "pattern": "*https://matrix.to/#/@alice:example.org*"
       }
     ],
     "actions" : [
       "notify",
       {
         "set_tweak": "sound",
         "value": "default"
       },
       {
         "set_tweak": "highlight"
       }
     ]
  }
  ```

  </details>

The last two proposals use a similar idea of attempting to find "pills" in the
`formatted_body`, this has some downsides though:

* It doesn't allow for hidden mentions, which can be useful in some situations.
* It does not work for event types other than `m.room.message`.

It also adds significant implementation complexity since the HTML messages must
now be parsed for notifications. This is expensive and introduces potential
security issues.

### Room version for backwards compatibility

Alternative backwards compatibility suggestions included using a new room version,
similar to [MSC3932](https://github.com/matrix-org/matrix-spec-proposals/pull/3932)
for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable

> room version before use, which would unnecessarily delay usage.

Reviewer 11: ```suggestion
room version before use, which would unnecessarily delay usage. Another MSC
can address this concern, such as in the extensible events series, if
desirable to be gated by a room version for a "clean slate" approach.
```

## Security considerations

> None foreseen.

Reviewer 11: ```suggestion
None not already described.
```

(abuse is a security concern)

## Unstable prefix

During development the following mapping will be used:

| What                | Stable            | Unstable                             |
|---------------------|-------------------|--------------------------------------|
| Event property      | `m.mentions`      | `org.matrix.msc3952.mentions`        |
| Push rule ID        | `.m.rule.*`       | `.org.matrix.msc3952.*`              |

If a client sees this rule available it can choose to apply the deprecation logic
discussed in the [backwards compatibility](#backwards-compatibility) section.

## Dependencies

> This depends on two MSCs which, at the time of writing, have not yet been accepted

Reviewer 11: ```suggestion
This depends on the following MSCs which, at the time of writing, have not yet been accepted
```

into the specification:

* [MSC3758](https://github.com/matrix-org/matrix-spec-proposals/pull/3758): Add `event_property_is` push rule condition kind
* [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873): event_match dotted keys
* [MSC3966](https://github.com/matrix-org/matrix-spec-proposals/pull/3966): `event_property_contains` push rule condition

> [^1]: It is [defined as](https://spec.matrix.org/v1.5/client-server-api/#mroommessage-msgtypes)

Reviewer 11: ```suggestion
----

[^1]: It is [defined as](https://spec.matrix.org/v1.5/client-server-api/#mroommessage-msgtypes)
```

Just to separate footnotes from dependencies a bit

Author: GitHub automatically adds a horizontal rule when rendering. Will double check this doesn't add 2.

Author: It adds a second horizontal rule when you do that, I'll add a comment just to separate them a bit.

```suggestion
<!-- Footnotes below: -->

[^1]: It is [defined as](https://spec.matrix.org/v1.5/client-server-api/#mroommessage-msgtypes)
```

# MSC3952: Intentional Mentions

Mentioning other users on Matrix is difficult -- it is not possible to know if
[mentioning a user by display name or Matrix ID](https://github.com/matrix-org/matrix-spec/issues/353)
will count as a mention, but is also too easy to mistakenly mention a user.

(Note that throughout this proposal "mention" is considered equivalent to a "ping"
or highlight notification.)

Some situations that result in unintentional mentions include:

* Replying to a message will re-issue pings from the initial message due to

>   [fallback replies](https://spec.matrix.org/v1.5/client-server-api/#fallbacks-for-rich-replies).

Author: This is particular bad if a user without the power to issue `@room` is replied to by someone who *can*, which will cause an `@room` mention.

# MSC3952: Intentional Mentions

Mentioning other users on Matrix is difficult -- it is not possible to know if
[mentioning a user by display name or Matrix ID](https://github.com/matrix-org/matrix-spec/issues/353)
will count as a mention, but is also too easy to mistakenly mention a user.

(Note that throughout this proposal "mention" is considered equivalent to a "ping"
or highlight notification.)

Some situations that result in unintentional mentions include:

* Replying to a message will re-issue pings from the initial message due to
  [fallback replies](https://spec.matrix.org/v1.5/client-server-api/#fallbacks-for-rich-replies).

>   * A user without the power level to send `@room` can abuse this by includin

Reviewer 19: ```suggestion
  * A user without the power level to send `@room` can abuse this by including
```

even be interpreted as a mention (see [issue 353](https://github.com/matrix-org/matrix-spec/issues/353)).
This results in some unexpected behavior and bugs:

* Matrix users use "leetspeak" when sending messages to avoid mentions (e.g.
  referring to M4tthew instead of Matthew).
* It is impossible to ping one out of multiple people with the same localpart
  (or display name).
* Since the relation between `body` and `formatted_body` is ill-defined and
  ["pills" are converted to display names](https://github.com/matrix-org/matrix-spec/issues/714),
  this can result in missed messages. [^1]

There are also some other related bugs:

* Matrix users will append emoji or other unique text in their display names to
  avoid unintentional pings.
* Bridging mentions is suboptimal since they [use display names](https://github.com/matrix-org/matrix-spec/issues/353#issuecomment-1055809364)
  as a workaround, e.g.:
  * It breaks the contract that bridges will not mutate the content of messages.
  * For some protocols, bridges need try to figure out if every message contains
    any of the possible nicknames of room members.
* If a user changes their display name in a room,

>   [they might not be mentioned unless the historical display name](https://github.com/matrix-org/matrix-spec/issues/353#issuecomment-1055809372)

Author: Maybe it was just a bug, I'm not sure. I think there's enough evidence regardless that this MSC is a change! We can remove it -- but I think the point still stands that evaluating these rules is a bit complicated due to historical state.

  is used while processing push rules.

## Background

Mentions are powered by two of the default push rules that search an event's
`content.body` property for the current user's display name
([`.m.rule.contains_display_name`](https://spec.matrix.org/v1.5/client-server-api/#default-override-rules))
or the localpart of their Matrix ID ([`.m.rule.contains_user_name`](https://spec.matrix.org/v1.5/client-server-api/#default-content-rules)).

There's also a [section about "user and room mentions"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
which defines that messages which mention the current user in the `formatted_body`
of the message should be colored differently:

> If the current user is mentioned in a message (either by a mention as defined
> in this module or by a push rule), the client should show that mention differently
> from other mentions, such as by using a red background color to signify to the
> user that they were mentioned.

## Proposal

> The existing push rules for mentions are deprecated and replaced with rules

Author: The end goal is that the deprecated ones will no longer exist.

There's an intermediate state (describe in more detail lower in ths MSC) where both exist, but the deprecated ones are disabled if the `m.mentions` property exists.

Any suggestion of how to reword this?

      ]
    },
    "m.relates_to": {
      "m.in_reply_to": {
          "event_id": "$initial_event"
      }
    }
  },
  // other fields as required by events
}
```

This signals that it is the *intention* of the sender to mention all of those people,
clients may wish to allow users to modify the list of people to include, e.g. to
"quote reply" as opposed to replying directly.

If a user wishes to be notified of *all replies* to their messages, other solutions
should be investigated, such as [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).
This would give more equal power to both senders and receivers of events.

### Impact on edits

Similarly to [replies](#impact-on-replies), users are notified of message edits
via the `.m.rule.contains_display_name` or the `.m.rule.contains_user_name` push
rule matching the [fallback content](https://spec.matrix.org/v1.6/client-server-api/#event-replacements).
Generally this is undesirable and users do not need to be notified for the same
message multiple times (e.g. if a user is fixing a typo). It is recommended that
clients include an empty `m.mentions` property when editing an event, *unless*

> the edit is significant or if additional users are mentioned.

Reviewer 19: Non blocking, but you could clarify between the `m.mentions` in the event itself and the one in the `m.new_content` block as it took me a couple of reads to realise they were two different things.

Author: I called it out in 4147590646cfd217f9d318a79c2cf85dd2b4afee that there are two copies. I think this is much clearer now!

  part).

Overall this proposal seems to be neutral or positive in the ability to combat
malicious behavior.

## Future extensions

### Combating abuse

Some ideas for combating abuse came from our discussion and research which are
worth sharing. These ideas are not a requirement for implementing this MSC, and
generally do not depend on it. (They could be implemented today with enough effort.)

It was recommended that clients could expose *why* an event has caused a notification
and give users inline tools to combat abuse. For example, a client might show a tooltip:

> The sender of the message (`@alice:example.org`) mentioned you in this event.
>
> Block `@alice:example.org` from sending you messages? `[Yes]` `[No]`

Additionally, if a user sending a message is about to mention many people it can
be useful to confirm whether they wish to do that (or prompt them to do an `@room`
mention instead).

Moderators may find tooling to quickly find messages which mention many users
useful in combating mention spammers. (Note that this should be made easier by
this MSC.)

A future MSC might wish to explore features for trusted contacts or soft-ignores

> to give users more control over who can generate notifications.

Reviewer 19: A bit of discussion on norm-setting would be interesting. That is, do we still expect clients to highlight the user's pill in a message red? What about in a message that includes their pill but doesn't mention them? Would we still expect user pills to be part of mention messages or do we see them being phased out in favour of just having the out-of-line intentional mentions ("x mentioned you in this event") that you discuss?

Reviewer 3: For the first part, to borrow the paradigm from discord; The @-pills would still be there, but they just wouldn't be highlighted, meaning they only exist to refer readers to the mentioned, but not to refer the mentioned to the message.

For the second, especially about "do we see them being phased out in favour of just having the out-of-line intentional mentions", I personally don't see this happening, as its still useful to refer users in messages, similar to "mentioning by name" works. Ultimately, I think that'll be up to social conventions, which means that changing between no pill-mentions (i.e. the "CC-like list" that the Potential for Abuse section mentions) or only-pill mentions is up to individual/group/cultural preference, meaning clients would have to deal with a mix of pill-mentions and non-pill-mentions in any message, it can be seen as a feature.

Reviewer 15: IMO, the out-of-line intentional ("no-pill") mentions still need to be rendered client-side somehow, otherwise the whole thing is going to become really confusing.

> That is, do we still expect clients to highlight the user's pill in a message red?

I would expect clients to do this *if and only if* the message contains an out-of-line intentional mention for the same user. Essentially, the presence of an out-of-line mention for a given user should activate special rendering of pills for that user. In cases where there is no out-of-line mention, IMO you shouldn't even be able to pillify, that is the pill would become ordinary text. Conversely, when there *is* an out-of-line mention but no pill, the client should decorate the entire message with something like "x thought you'd be interested in this message" or something along those lines.

Clients might want to simplify and not expose this last case to the user, requiring the user to produce a pill in order to even send out an out-of-line intentional mention. (Of course, you'd still be able to produce such events manually).

Reviewer 3: > IMO you shouldn't even be able to pillify, that is the pill would become ordinary text.

there might be a reason why the original poster would have chosen to turn a person's mention into a pill, instead of "just plain text", maybe to point the other user(s) in the chat to "this specific user" instead of "this person we know by this name", which'd they have to lookup and search (in the room's member list), or simply know someone by that alias beforehand, to know who the OP is talking about.

> Clients might want to simplify and not expose this last case to the user, requiring the user to produce a pill in order to even send out an out-of-line intentional mention.

Agreed, I see this as falling in line with matrix's client simplicity preference/option, selectively pruning some advanced options to simplify the interface for their users.

Reviewer 15: > there might be a reason why the original poster would have chosen to turn a person's mention into a pill, instead of "just plain text", maybe to point the other user(s) in the chat to "this specific user" instead of "this person we know by this name", which'd they have to lookup and search (in the room's member list), or simply know someone by that alias beforehand, to know who the OP is talking about.

That's a fair (and convincing) point which I didn't think of.

Though one related thing that I would personally find annoying is if someone mentioned me in this way but my client provided no ability to notify me about it.

Author: Hopefully b9756431eecd2edebe3d4b71da3b0d703a115c7d clarifies this a bit.

Reviewer 19: Thanks, I think that does help.

On the discussion, I'm afraid I'm really only interested in the text that goes into the MSC rather than what people think about it. I would prefer to keep discussion to what we want the MSC to say.

for extensible events. This does not seem like a good fit since room versions are
not usually interested in non-state events. It would additionally require a stable
room version before use, which would unnecessarily delay usage. Another MSC
can address this concern, such as in the extensible events series, if
desirable to be gated by a room version for a "clean slate" approach.

## Security considerations

None not already described.

## Unstable prefix

During development the following mapping will be used:

| What                | Stable            | Unstable                             |
|---------------------|-------------------|--------------------------------------|
| Event property      | `m.mentions`      | `org.matrix.msc3952.mentions`        |
| Push rule ID        | `.m.rule.*`       | `.org.matrix.msc3952.*`              |

The server will include the `org.matrix.msc3952_intentional_mentions` flag in the
`unstable_features` array of the `/versions` endpoint. If a client sees this flag
it can choose to apply the deprecation logic discussed in the
[backwards compatibility](#backwards-compatibility) section.

## Dependencies

> This depends on the following MSCs which, at the time of writing, have not yet been accepted
> into the specification:

Reviewer 11: ```suggestion
This depends on the following MSCs:
```

* One inside the `m.new_content` property, which should contain the full list of
  mentioned users in any version of the event.

It is recommended that clients use an empty top-level `m.mentions` property when
editing an event, *unless* the edit is significant or if additional users are
mentioned.

For example, if there is an event:

```json5
{
  "sender": "@dan:example.org",
  "event_id": "$initial_event",
  "content": {
    "body": "Helo Alice!",
    "m.mentions": {
      "user_ids": ["@alice:example.org"]
    }
  },
  // other fields as required by events
}
```

And an edit after realizing that Bob is also in the room:

```json5
{
  "content": {
    "body": "* Hello Alice & Bob!",
    "m.mentions": {
      "user_ids": [
        // Include only the newly mentioned user.
        "@bob:example.org"
      ]
    },
    "m.new_content": {
      "body": "Hello Alice & Bob!",
      "m.mentions": {
        "user_ids": [
          // Include all mentioned users.
          "@alice:example.org",
          "@bob:example.org"
        ]
      },
    },
    "m.relates_to": {
      "rel_type": "m.replace",
      "event_id": "$initial_event"
    }
  },
  // other fields as required by events
}

> ```

Reviewer 13: If I'm understanding this correctly, it may be worth calling out that:

- If a client highlight a message visually (e.g. by colouring it red), then it should look at the `m.mentions` under `m.new_content` for edited messages.  Otherwise, in the example above, Alice would not see the message as red, even though the intent was for her to be mentioned.
- If a client uses some sort of notification sound, it should look at the `m.mentions` directly under `content` and not ding for Alice (if it already dinged for the original message).

Author: It is probably worth calling out that your second bullet point is really just "process push rules on the event as normal".

Author: 1d1c8f3159134248cc62e111ad9be9b09a403560 clarifies this a bit.

This does give an odd experience if you were mentioned on a previous edit and then removed. The message wouldn't be highlighted, but the user may be confused at why they got a notification. I think this is a separate problem to solve honestly.

property would not include the removed user IDs (you cannot cancel the notification from
the previous event) or any previously notified users, and the removed user would also be
removed from the `m.new_content` proprerty's copy  of `m.mentions`.

This should limit duplicate, unnecessary notifications for users. If a user wishes
to receive notifications for edits of events they were mentioned in then they
could setup a push rule for the `content.m\\.new_content.m\\.mentions` property
or potentially leverage [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).

### Impact on bridging

This should strengthen the bridging contract as it will allow them to stop
mutating the content of messages during the bridging process (by directly
listing any mentioned users in the `m.mentions` property).

This should reduce the impedance mismatch when bridging protocols which similarly
list out users to mention (the bridge should simply be able to map from the
remote user ID to the bridged user ID)[^4], but will not help in cases where the
mention information is embedded into the text content (as today's Matrix messages
are).[^5]

> 

Author: You mean going from Matrix -> Bridge? Correct. I can reword this a bit to be clearer. (Although from Bridge -> Matrix we shouldn't need to do that anymore.)

for the client to handle this feature and it must be handled on the server, unfortunately
this would not be possible with the current proposal since the `m.mentions`
property is encrypted (and the server cannot act on it).

Solving this problem is left to a future MSC.

### Pillifying `@room`

Some clients attempt to create a "pill" out of `@room` mentions, but this is not
a requirement of the Matrix specification. The current [user and rooms mentions](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions)
section could be expanded for this situation.

### Extensible events

Handling of this property in [MSC1767](https://github.com/matrix-org/matrix-doc/pull/1767)-style
extensible events is deliberately left for a future MSC to address, if needed.

### Role mentions

It is possible to add additional properties to the `m.mentions` object, e.g. a foreseeable
usecase would be a `roles` property which could include values such as `admins` or
`mods`. Defining this in detail is left to a future MSC.

### Cancelling notifications

It maybe useful for a future MSC to investigate cancelling notifications if a

> user's mentioned is removed while [editing events](#impact-on-edits). This could
> be quite difficult as it is unclear if the mentioned user has already receive

Reviewer 13: ```suggestion
user's mention is removed while [editing events](#impact-on-edits). This could
be quite difficult as it is unclear if the mentioned user has already received
```

can address this concern, such as in the extensible events series, if
desirable to be gated by a room version for a "clean slate" approach.

## Security considerations

None not already described.

## Unstable prefix

During development the following mapping will be used:

| What                | Stable            | Unstable                             |
|---------------------|-------------------|--------------------------------------|
| Event property      | `m.mentions`      | `org.matrix.msc3952.mentions`        |
| Push rule ID        | `.m.rule.*`       | `.org.matrix.msc3952.*`              |

The server will include the `org.matrix.msc3952_intentional_mentions` flag in the
`unstable_features` array of the `/versions` endpoint. If a client sees this flag
it can choose to apply the deprecation logic discussed in the
[backwards compatibility](#backwards-compatibility) section.

## Dependencies

This depends on the following MSCs:

* [MSC3758](https://github.com/matrix-org/matrix-spec-proposals/pull/3758): Add `event_property_is` push rule condition kind
* [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873): event_match dotted keys

> * [MSC3966](https://github.com/matrix-org/matrix-spec-proposals/pull/3966): `event_property_contains` push rule condition

Reviewer 13: For reference, these three MSCs have already been accepted.

Author: Is it clearer if I just remove them now as they've all been accepted?

Reviewer 13: No, I think it's better to keep it around for historical documentation.  I was just leaving a note for the sake of other reviewers to indicate that these MSCs aren't blocking this one.

It is valid to include both the `user_ids` and `room` properties.

It is recommended that homeservers reject locally created events with an invalid
`m.mentions` property with an error with a status code of `400` and an errcode of
`M_INVALID_PARAM`.

Clients add a Matrix ID to the `user_ids` array whenever composing a message which
includes an intentional mention, such as a ["pill"](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions).
Clients set the `room` value to `true` when making a room-wide announcement. Clients
should also set these values at other times when it is obvious the user intends to explicitly
mention a user.[^3]

The `m.mentions` property is part of the plaintext event body and should be encrypted
into the ciphertext for encrypted events.

### New push rules

Two new default push rule are added:

The `.m.rule.is_user_mention` override push rule would appear directly
before the `.m.rule.contains_display_name` push rule:

```json
{
    "rule_id": ".m.rule.is_user_mention",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "event_property_contains",

>             "key": "content.m\\.mentions.user_ids",

Author: We [talked about this in #matrix-spec](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$WJgu_9SclbeiRCc8QKVU9PABZqwQ9YXWxSZzL7JJwRo?via=matrix.org&via=libera.chat&via=element.io), see 811c0df5ecbc65333ae99ec3f825ce1175d9c885 for the result.

tl;dr this is JSON, so the double backslash is correct (but actually represents a single character -- a backslash).

`.m.rule.contains_user_name` push rule matching the
[rich reply fallback](https://spec.matrix.org/v1.6/client-server-api/#fallbacks-for-rich-replies).
Unfortunately these push rules will be disabled for events  which contain the
`m.mentions` property, i.e. all newly created events (see
[above](#backwards-compatibility)). It is proposed that clients should include
the sender of the event being replied to as well as any users (except themself)
mentioned in that event in the new event's `m.mentions` property. The `room`
property should not be copied over.

For example, if there is an event:

```json5
{
  "sender": "@dan:example.org",
  "event_id": "$initial_event",
  "content": {
    "body": "Alice: Have you heard from Bob?",
    "m.mentions": {
      "user_ids": ["@alice:example.org", "@bob:example.org"]
    }
  },
  // other fields as required by events
}
```

And a reply from Alice:

```json5
{
  "content": {
    "body": "> <@dan:example.org> Alice: Have you heard from Bob?\n\nNo, but I saw him with Charlie earlier.",
    "m.mentions": {
      "user_ids": [
        // Include the sender of $initial_event.
        "@dan:example.org",
        // The users mentioned, minus yourself.

>         "@bob:example.org",

Author: > I'd expect the same behaviour from Matrix clients, so it may be worth having the example reflect that behaviour (unless you disagree?).

I'm attempting to *not* change today's behavior (compared to the reply-fallbacks we have now) when sending replies.

> Otherwise, I see that the point of this section is to communicate that clients *could* include whichever users they would like. I still worry that clients will just copy the example in the spec.

Yes clients can choose to do different things here, that might appear flakey if clients do different things though! I think ideally it would be consistent across clients?

>         "@bob:example.org",

Reviewer 5: > that might appear flakey if clients do different things though

Do you mean during the transition period? Clients are already going to be doing different things during that period anyways (some sending unintended mentions, whereas others are not).

> I think ideally it would be consistent across clients?

Indeed! Though I fear some clients may copy the example and mention other users during a reply, whereas other clients would choose not to do so.

I only raise this point as I've seen past cases of clients blindly using fields from examples in the spec *which aren't actually specced*. Those fields originally being c/p'd from a client implementation in the wild without being properly vetted.

> I'm attempting to not change today's behavior (compared to the reply-fallbacks we have now) when sending replies.

I see that - but why not? We're fixing so many other behaviour around mentions in this MSC already.

>         "@bob:example.org",

Reviewer 5: @clokep and I talked this over in DMs, which has resulted in https://github.com/matrix-org/matrix-spec-proposals/pull/3952/commits/63592e7f63a2f007af91bf6739463a40a3c9c791. Which I'm happy with as it clarifies that different use cases ought to consider desired semantics for this behaviour.

        // New mentions, as normal.
        "@charlie:example.org"
      ]
    },
    "m.relates_to": {
      "m.in_reply_to": {
          "event_id": "$initial_event"
      }
    }
  },
  // other fields as required by events
}
```

This signals that it is the *intention* of the sender to mention all of those people,
clients may wish to allow users to modify the list of people to include, e.g. to
"quote reply" as opposed to replying directly.

If a user wishes to be notified of *all replies* to their messages, other solutions
should be investigated, such as [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).
This would give more equal power to both senders and receivers of events.

### Impact on edits

Similarly to [replies](#impact-on-replies), users are notified of message edits
via the `.m.rule.contains_display_name` or the `.m.rule.contains_user_name` push
rule matching the [fallback content](https://spec.matrix.org/v1.6/client-server-api/#event-replacements).
Generally this is undesirable and users do not need to be notified for the same
message multiple times (e.g. if a user is fixing a typo).

> Edited events end up with two `m.mentions` properties:

Reviewer 14: the *edited* event doesn't get a second `m.mention` property? That's in the replacement, not the original event (either before or after [applying `m.new_content`](https://spec.matrix.org/v1.6/client-server-api/#applying-mnew_content)). I think you mean:

```suggestion
Replacement events may have `m.mentions` properties in two locations:
```

* One at the top-level of the `content`, which should contain any users to mention
 *for this edit*.
* One inside the `m.new_content` property, which should contain the full list of
  mentioned users in any version of the event.

It is recommended that clients use an empty top-level `m.mentions` property when
editing an event, *unless* the edit is significant or if additional users are
mentioned.

For example, if there is an event:

```json5
{
  "sender": "@dan:example.org",
  "event_id": "$initial_event",
  "content": {

>     "body": "Helo Alice!",

Author: This is the typo being fixed below. 😄

>     "body": "Helo Alice!",

Author: Or maybe it used to be and I changed the example?

malicious behavior.

## Future extensions

### Combating abuse

Some ideas for combating abuse came from our discussion and research which are
worth sharing. These ideas are not a requirement for implementing this MSC, and
generally do not depend on it. (They could be implemented today with enough effort.)

It was recommended that clients could expose *why* an event has caused a notification
and give users inline tools to combat abuse. For example, a client might show a tooltip:

> The sender of the message (`@alice:example.org`) mentioned you in this event.
>
> Block `@alice:example.org` from sending you messages? `[Yes]` `[No]`

Additionally, if a user sending a message is about to mention many people it can
be useful to confirm whether they wish to do that (or prompt them to do an `@room`
mention instead).

Moderators may find tooling to quickly find messages which mention many users
useful in combating mention spammers. (Note that this should be made easier by
this MSC.)

A future MSC might wish to explore features for trusted contacts or soft-ignores
to give users more control over who can generate notifications.

> ### Muted except for mentions push rules

Reviewer 18: @erikjohnston FYI clokep prepared https://github.com/matrix-org/matrix-spec-proposals/pull/3996

> ### Muted except for mentions push rules

Author: 147ff7803ca927e8ff132a34125ac18c3766a1e7 attempts to clarify.

          "event_id": "$initial_event"
      }
    }
  },
  // other fields as required by events
}
```

This signals that it is the *intention* of the sender to mention all of those people,
clients may wish to allow users to modify the list of people to include, e.g. to
"quote reply" as opposed to replying directly.

If a user wishes to be notified of *all replies* to their messages, other solutions
should be investigated, such as [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).
This would give more equal power to both senders and receivers of events.

### Impact on edits

Similarly to [replies](#impact-on-replies), users are notified of message edits
via the `.m.rule.contains_display_name` or the `.m.rule.contains_user_name` push
rule matching the [fallback content](https://spec.matrix.org/v1.6/client-server-api/#event-replacements).
Generally this is undesirable and users do not need to be notified for the same
message multiple times (e.g. if a user is fixing a typo).

Replacement events may have `m.mentions` properties in two locations:

* One at the top-level of the `content`, which should contain any users to mention
 *for this edit*.

> * One inside the `m.new_content` property, which should contain the full list of
>   mentioned users in any version of the event.

Reviewer 5: I wrote a comment about what would happen if a user mention was removed in the edit... before reading on :)

```suggestion
* One inside the `m.new_content` property, which should contain the full list of
  mentioned users in any version of the event, unless a mention is removed
  (see below).
```

It is recommended that clients use an empty top-level `m.mentions` property when
editing an event, *unless* the edit is significant or if additional users are
mentioned.

For example, if there is an event:

```json5
{
  "sender": "@dan:example.org",
  "event_id": "$initial_event",
  "content": {
    "body": "Helo Alice!",
    "m.mentions": {
      "user_ids": ["@alice:example.org"]
    }
  },
  // other fields as required by events
}
```

And an edit after realizing that Bob is also in the room:

```json5
{
  "content": {
    "body": "* Hello Alice & Bob!",
    "m.mentions": {
      "user_ids": [
        // Include only the newly mentioned user.
        "@bob:example.org"
      ]
    },
    "m.new_content": {
      "body": "Hello Alice & Bob!",
      "m.mentions": {
        "user_ids": [
          // Include all mentioned users.
          "@alice:example.org",
          "@bob:example.org"
        ]
      },
    },
    "m.relates_to": {
      "rel_type": "m.replace",
      "event_id": "$initial_event"
    }
  },
  // other fields as required by events
}
```

> Mentions can also be removed as part of an edit. In this case top-level `m.mentions`

Reviewer 5: ```suggestion
Mentions can also be removed as part of an edit. In this case, the top-level `m.mentions`
```

property would not include the removed user IDs (you cannot cancel the notification from
the previous event) or any previously notified users, and the removed user would also be
removed from the `m.new_content` proprerty's copy  of `m.mentions`.

This should limit duplicate, unnecessary notifications for users. If a user wishes
to receive notifications for edits of events they were mentioned in then they

> could setup a push rule for the `content.m\\.new_content.m\\.mentions` property

Reviewer 5: likewise here.

```suggestion
could setup a push rule for the `content.m\.new_content.m\.mentions` property
```

or potentially leverage [MSC3664](https://github.com/matrix-org/matrix-spec-proposals/pull/3664).

This implies that:

* If a client highlights a message visually (e.g. by coloring it red), then it
  should look at the `m.mentions` under `m.new_content` for edited messages.
  Otherwise, in the example above, Alice would not see the message as red, even
  though the intent was for her to be mentioned.
* Any sort of processing of push rules, e.g. to display a notification (sound,
  toast, push notification), should occur without any special rules. I.e. the
  `.m.rule.is_user_mention` and `.m.rule.is_room_mention` should look at the
  `m.mentions` directly under `content` and not match for Alice.

### Impact on bridging

For protocols with a similar mechanism for listing mentioned users this should
strengthen the bridging contract as it enables bridges to stop mutating the
content of messages. The bridge should be able to map from the remote user ID
to the bridged user ID and include that in the `m.mentions` property of the
Matrix event & the proper field in the bridged protocol[^4].

> For bridged protocols that do not have this mechanism than the bridge will only

Reviewer 5: ```suggestion
For bridged protocols that do not have this mechanism, the bridge will only
```

mentioned Matrix IDs to be part of the displayed text. This is only a limitation
for current clients: mentions could be exposed in the user interface directly.
For example, a de-emphasized "notified" list could be shown on messages, similar
to CCing users on an e-mail.

Although not including mentions in the displayed text could be used as an abuse
vector, it does not enable additional malicious behavior than what is possible
today. From discussions and research while writing this MSC there are moderation
benefits to using a separate property for mentions:

* The number of mentions is trivially limited by moderation tooling, e.g. it may
  be appropriate for a community room to only allow 10 mentions. Events not abiding
  by this could be rejected automatically (or users could be banned automatically).
* Various forms of "mention bombing" are no longer possible.
* It is simpler to collect metrics on how mentions are being used (it is no longer
  necessary to process the textual `body` for every user's display name and local
  part).

Overall this proposal seems to be neutral or positive in the ability to combat
malicious behavior.

### Encrypted mentions & `/notifications`

> Previous version of this proposal (and the alternative [MSC1796](https://github.com/matrix-org/matrix-spec-proposals/pull/1796))

Reviewer 5: ```suggestion
A previous version of this proposal (and the alternative [MSC1796](https://github.com/matrix-org/matrix-spec-proposals/pull/1796))
```

        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "highlight"
        }
    ]
}
```

An example event matching both `.m.rule.is_user_mention` (for `@alice:example.org`)
and `.m.rule.is_room_mention` is provided below:

```json
{
    "content": {
        "body": "This is an example mention @alice:example.org",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>This is an example mention</b> <a href='https://matrix.to/#/@alice:example.org'>Alice</a>",
        "msgtype": "m.text",
        "m.mentions": {
            "user_ids": ["@alice:example.org"],
            "room": true
        }
    },
    "event_id": "$143273582443PhrSn:example.org",
    "origin_server_ts": 1432735824653,
    "room_id": "!somewhere:over.the.rainbow",
    "sender": "@example:example.org",
    "type": "m.room.message",
    "unsigned": {
        "age": 1234
    }
}
```

### Client behavior

The overall user experience is not modified, beyond improving explicitness and
reducing unintended mentions.

For example, it is common that a client will show an event with a mention in a
different color (and denote the current user's "pill", as a way of showing the

> user *why* they were mentioned). This behavior is unchanged.

Author: > I wonder if we should explicitly discourage receiving clients from pillifying display names of mentioned users. I admit it's a somewhat corner case because well-behaved sending clients are supposed to use MXIDs and not display names in the message body; it's more to protect overzealous client authors from a pitfall.

I thought the [current spec](https://spec.matrix.org/v1.6/client-server-api/#user-and-room-mentions) was a bit clearer on this then it is, but yeah the intention is that only matrix.to links should be pillified on receiving. (I don't think this MSC is really changing that though FTR.)

> user *why* they were mentioned). This behavior is unchanged.

Reviewer 20: I admit I didn't read the msc much but I assume you also mean pills for matrix: URLs and not just matrix.to? Feels like a step back if only matrix.to would be pillified (apart from it being a centralised service still)

> user *why* they were mentioned). This behavior is unchanged.

Author: > I admit I didn't read the msc much but I assume you also mean pills for matrix: URLs and not just matrix.to? Feels like a step back if only matrix.to would be pillified (apart from it being a centralised service still)

The spec currently only defines them for matrix.to IIRC, I suspect this was an oversight when matrix: URIs were accept though?

This MSC makes *no changes about how/when pillifying is done*. It does say that if you pillify than the user should be mentioned explicitly in the event metadata.

> user *why* they were mentioned). This behavior is unchanged.

Reviewer 15: > The spec currently only defines them for matrix.to IIRC, I suspect this was an oversight when matrix: URIs were accept though?

It seems to me that this is not even the case. The spec [talks about including `Matrix URIs`](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions) in the event to make a mention, which it defines as either a matrix.to link or a `matrix:` URI. The example given uses a matrix.to anchor, but it is just an example.

> user *why* they were mentioned). This behavior is unchanged.

Author: Ah, good-catch. The [Matrix URI](https://spec.matrix.org/v1.5/appendices/#uris) section seems to talk about both!

> user *why* they were mentioned). This behavior is unchanged.

Author: Anyway -- I don't think we're making any particular change here, so I'm going to resolve this.

that both sets of push rules will need to be supported for a period of time, but
at worst case should simply result in the current behavior (documented in the preamble).

If users wish to continue to be notified of messages containing their display name
it is recommended that clients create a specific keyword rule for this, e.g. a
`content` rule of the form:

```json
{
  "actions": [
    "notify",
    {
      "set_tweak": "sound",
      "value": "default"
    },
    {
      "set_tweak": "highlight"
    }
  ],
  "pattern": "alice",
  "rule_id": "alice",
  "enabled": true
}
```

### Impact on replies

Users are notified of replies via the `.m.rule.contains_display_name` or the
`.m.rule.contains_user_name` push rule matching the
[rich reply fallback](https://spec.matrix.org/v1.6/client-server-api/#fallbacks-for-rich-replies).
Unfortunately these push rules will be disabled for events  which contain the
`m.mentions` property, i.e. all newly created events (see
[above](#backwards-compatibility)). Clients should include the sender of the event
being replied to as well as any mentioned users in that event (excluding yourself)

> in the new event's `m.mentions` property. The `room` property should not be copied over.

Author: I don't know the procedure of whether to commit this or not now that this has completed FCP. I'm fine with the change to the text though.

@turt2live do you have any thoughts?

> in the new event's `m.mentions` property. The `room` property should not be copied over.

Reviewer 13: The comment was made before FCP was over, and seems like a minor enough change.  So if you agree with it, I'd say that it's fine to commit it.
